{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cb4c5e28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "siteuse_rea 235622\n",
      "numibd 202906\n",
      "done gene\n",
      "done mu\n",
      "siteuse_sim 73690\n",
      "numibd 49092\n",
      "-----------------------------------\n",
      "current_pop [3653.8461538461524]\n",
      "[2004887] 4373507\n",
      "[228] 1246\n",
      "[252913.46520818057] 61191.03345391462\n",
      "[574513.2222592189] 186687.3637054744\n",
      "153.24276328086853 0\n",
      "-----------------------------------\n",
      "done gene\n",
      "done mu\n",
      "siteuse_sim 71396\n",
      "numibd 49062\n",
      "-----------------------------------\n",
      "current_pop [3653.8461538461524, 3653.8461538461524]\n",
      "[2004887, 1845658] 4373507\n",
      "[228, 164] 1246\n",
      "[252913.46520818057, 253068.11450817334] 61191.03345391462\n",
      "[574513.2222592189, 570253.9735537305] 186687.3637054744\n",
      "300.6347608566284 1\n",
      "-----------------------------------\n",
      "done gene\n",
      "done mu\n",
      "siteuse_sim 80006\n",
      "numibd 56667\n",
      "-----------------------------------\n",
      "current_pop [3653.8461538461524, 3653.8461538461524, 7454.697880374698]\n",
      "[2004887, 1845658, 1978240] 4373507\n",
      "[228, 164, 124] 1246\n",
      "[252913.46520818057, 253068.11450817334, 219105.08468773714] 61191.03345391462\n",
      "[574513.2222592189, 570253.9735537305, 493037.996742682] 186687.3637054744\n",
      "469.81083059310913 2\n",
      "-----------------------------------\n",
      "done gene\n",
      "done mu\n",
      "siteuse_sim 83000\n",
      "numibd 57490\n",
      "-----------------------------------\n",
      "current_pop [3653.8461538461524, 3653.8461538461524, 7454.697880374698, 7454.697880374698]\n",
      "[2004887, 1845658, 1978240, 2045256] 4373507\n",
      "[228, 164, 124, 140] 1246\n",
      "[252913.46520818057, 253068.11450817334, 219105.08468773714, 215968.47858758044] 61191.03345391462\n",
      "[574513.2222592189, 570253.9735537305, 493037.996742682, 488484.6394612719] 186687.3637054744\n",
      "638.6174886226654 3\n",
      "-----------------------------------\n",
      "done gene\n",
      "done mu\n",
      "siteuse_sim 70440\n",
      "numibd 49625\n",
      "-----------------------------------\n",
      "current_pop [3653.8461538461524, 3653.8461538461524, 7454.697880374698, 7454.697880374698, 3653.846153846156]\n",
      "[2004887, 1845658, 1978240, 2045256, 1793351] 4373507\n",
      "[228, 164, 124, 140, 114] 1246\n",
      "[252913.46520818057, 253068.11450817334, 219105.08468773714, 215968.47858758044, 250197.03443828714] 61191.03345391462\n",
      "[574513.2222592189, 570253.9735537305, 493037.996742682, 488484.6394612719, 559010.3565838945] 186687.3637054744\n",
      "789.9684710502625 4\n",
      "-----------------------------------\n",
      "done gene\n",
      "done mu\n",
      "siteuse_sim 70658\n",
      "numibd 50684\n",
      "-----------------------------------\n",
      "current_pop [3653.8461538461524, 3653.8461538461524, 7454.697880374698, 7454.697880374698, 3653.846153846156, 3653.846153846156]\n",
      "[2004887, 1845658, 1978240, 2045256, 1793351, 1826841] 4373507\n",
      "[228, 164, 124, 140, 114, 144] 1246\n",
      "[252913.46520818057, 253068.11450817334, 219105.08468773714, 215968.47858758044, 250197.03443828714, 244969.37562149792] 61191.03345391462\n",
      "[574513.2222592189, 570253.9735537305, 493037.996742682, 488484.6394612719, 559010.3565838945, 555069.6204547792] 186687.3637054744\n",
      "944.446914434433 5\n",
      "-----------------------------------\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import statistics\n",
    "import msprime\n",
    "from IPython.display import SVG, display\n",
    "import tsinfer\n",
    "import scipy\n",
    "import math\n",
    "import numpy\n",
    "import tskit\n",
    "import io\n",
    "import builtins\n",
    "import sys\n",
    "from tqdm.notebook import tqdm\n",
    "from tskit import MISSING_DATA\n",
    "import pickle\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import cvxopt as opt\n",
    "from cvxopt import matrix, spmatrix, sparse\n",
    "from cvxopt.solvers import qp, options\n",
    "from cvxopt import blas\n",
    "from cvxopt import spdiag\n",
    "import statsmodels\n",
    "from statsmodels.nonparametric.kernel_regression import KernelReg\n",
    "from scipy.signal import savgol_filter\n",
    "import time\n",
    "import bisect\n",
    "\n",
    "\n",
    "def find_s(elem, sorted_list):\n",
    "    'Locate the leftmost value exactly equal to x'\n",
    "    i = bisect.bisect_left(sorted_list, elem)\n",
    "    if i != len(sorted_list) and sorted_list[i] == elem:\n",
    "        return i\n",
    "\n",
    "\n",
    "def find_s_se(elem, sorted_list):\n",
    "    'Locate the leftmost value smaller or equal to x'\n",
    "    i = bisect.bisect_right(sorted_list, elem)\n",
    "    if i != 0:\n",
    "        return i-1\n",
    "    else:\n",
    "        print(\"error\")\n",
    "\n",
    "\n",
    "def find_s_s(elem, sorted_list):\n",
    "    'Locate the leftmost value smaller or equal to x'\n",
    "    i = bisect.bisect_left(sorted_list, elem)\n",
    "    if i != 0:\n",
    "        return i\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "    \n",
    "file = open(\"recombination_map.dat\",\"rb\")\n",
    "rmap = pickle.load(file)\n",
    "file.close\n",
    "\n",
    "sam=tsinfer.load(\"tgp_chr20.samples\")\n",
    "\n",
    "LWKlist = []\n",
    "for i in range(len(sam.individuals_population)):\n",
    "    if sam.individuals_population[i]==11:\n",
    "        LWKlist.append(i)\n",
    "newdata = sam.subset(individuals=LWKlist)\n",
    "\n",
    "\n",
    "        \n",
    "seq_len = 63025522\n",
    "sam_size = 99\n",
    "r = rmap\n",
    "mu = 1.375*10**(-8)\n",
    "# gene_conver = 10**(-8)\n",
    "# track_len = 100\n",
    "# epsilon = 0.0001\n",
    "totaliter=1\n",
    "\n",
    "simtime = 100\n",
    "# priormu = [1.0*10**(-8), 2.0*10**(-8)]\n",
    "# prioreps = [0, 0.0002]\n",
    "# accept_threshold = [1000,400]\n",
    "freqconsider = [1,40]\n",
    "priorinipop = [60000, 5000]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "inipopsimlist = []\n",
    "\n",
    "summary_sta1_simlist = []\n",
    "summary_sta2_simlist = []\n",
    "summary_sta3_simlist = []\n",
    "summary_sta4_simlist = []\n",
    "\n",
    "\n",
    "for iter in range(totaliter):\n",
    "\n",
    "    sitefreq = numpy.sum(newdata.sites_genotypes,1)\n",
    "    siteuse =  numpy.where( (sitefreq>freqconsider[0] ) * ( sitefreq<freqconsider[1] ) )[0]  \n",
    "    print(\"siteuse_rea\",len(siteuse))\n",
    "\n",
    "    datause = newdata.subset(sites = siteuse)\n",
    "\n",
    "    inferred_ts = tsinfer.infer(datause,path_compression=False)\n",
    "\n",
    "\n",
    "\n",
    "    mts = inferred_ts\n",
    "\n",
    "    ###############################################################################\n",
    "    msites = mts.tables.mutations.site\n",
    "    sitepos = mts.tables.sites.position\n",
    "    mpos = []\n",
    "    for i in range(len(msites)):\n",
    "        cursite = msites[i]\n",
    "        mpos.append(sitepos[cursite])\n",
    "    mpos = numpy.array(mpos)\n",
    "    mnodes = mts.tables.mutations.node\n",
    "    ec = mts.tables.edges.child\n",
    "    ep = mts.tables.edges.parent\n",
    "    el = mts.tables.edges.left\n",
    "    er = mts.tables.edges.right\n",
    "\n",
    "\n",
    "    leftmost = min(el)\n",
    "    rightmost = max(er)\n",
    "    for i in range(len(el)):\n",
    "        if el[i] == leftmost:\n",
    "            el[i] = 0\n",
    "        if er[i] == rightmost:\n",
    "            er[i] = seq_len\n",
    "\n",
    "\n",
    "    if len(mpos) == 0:\n",
    "        mpos_sn = []\n",
    "        mnodes_sn = []\n",
    "    else:\n",
    "        mnp_sn = numpy.transpose(sorted(numpy.transpose(numpy.array([mnodes,mpos])).tolist()))\n",
    "        mnodes_sn = mnp_sn[0]\n",
    "        mpos_sn = mnp_sn[1]\n",
    "\n",
    "    eclrp_sc = numpy.transpose(sorted(numpy.transpose(numpy.array([ec,el,er,ep])).tolist()))\n",
    "    ec_sc = eclrp_sc[0]\n",
    "    el_sc = eclrp_sc[1]\n",
    "    er_sc = eclrp_sc[2]\n",
    "    ep_sc = eclrp_sc[3]\n",
    "    enm_sc = [[] for i in range(len(ec_sc))]\n",
    "\n",
    "    curedge = 0\n",
    "    for i in range(len(mnodes_sn)):\n",
    "        curnodemu = mnodes_sn[i] \n",
    "        cursitemu = mpos_sn[i]\n",
    "        for j in range(curedge,len(ec_sc)):\n",
    "            if curnodemu == ec_sc[j]:\n",
    "\n",
    "                if er_sc[j] > cursitemu and el_sc[j] <= cursitemu:\n",
    "\n",
    "                    enm_sc[j].append(cursitemu)  \n",
    "\n",
    "                    curedge = j\n",
    "                    break\n",
    "                if el_sc[j] > cursitemu:\n",
    "                    curedge = j\n",
    "                    break\n",
    "\n",
    "            if ec_sc[j] > mnodes_sn[i] or j == len(ec_sc)-1:\n",
    "                #print(\"missing mutations\")\n",
    "                curedge = j\n",
    "                break\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ###########################################################################\n",
    "    ###########################################################################\n",
    "    ##########################################################################\n",
    "\n",
    "    clocate = []\n",
    "    curc = -1\n",
    "    for i in range(len(ec_sc)):\n",
    "        while ec_sc[i] != curc:\n",
    "            curc=curc+1\n",
    "            clocate.append(i)\n",
    "    clocate.append(len(ec_sc))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    abdlist = []\n",
    "\n",
    "    for i in range(0,2*sam_size-1):\n",
    "        for j in range(i+1,i+2):\n",
    "            canabd = []\n",
    "            curleft = 0\n",
    "            curright =seq_len\n",
    "            #start = 0\n",
    "            while curleft != seq_len:\n",
    "                curp1 = i\n",
    "                curp2 = j \n",
    "                curmed = []\n",
    "                while curp1 != curp2:\n",
    "                    if curp1 < curp2:\n",
    "                        begin1 = clocate[curp1]\n",
    "                        end = clocate[curp1+1]\n",
    "                        s_range = el_sc[begin1:end]\n",
    "                        begin2 = find_s_se(curleft,s_range)\n",
    "                        for k in range(begin1+begin2,end):\n",
    "                            if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                curp1 = int(ep_sc[k])\n",
    "                                curoldr = curright\n",
    "                                curright = min(er_sc[k],curright)\n",
    "\n",
    "                                curmed.append(k)\n",
    "\n",
    "                                break\n",
    "\n",
    "                    else:\n",
    "                        begin1 = clocate[curp2]\n",
    "                        end = clocate[curp2+1]\n",
    "                        s_range = el_sc[begin1:end]\n",
    "                        begin2 = find_s_se(curleft,s_range)\n",
    "                        for k in range(begin1+begin2,end):\n",
    "                            if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                curp2 = int(ep_sc[k])\n",
    "                                curoldr = curright\n",
    "                                curright = min(er_sc[k],curright) \n",
    "\n",
    "                                curmed.append(k)\n",
    "\n",
    "                                break\n",
    "\n",
    "                canmution = []\n",
    "                for k in range(len(curmed)):\n",
    "                    edgemu = enm_sc[curmed[k]]\n",
    "                    startt = find_s_s(curleft,edgemu)\n",
    "                    endd = find_s_s(curright,edgemu)                    \n",
    "                    canmution = canmution + edgemu[startt:endd] \n",
    "\n",
    "                canabd.append([i,j,curleft,curright,curp1,canmution])\n",
    "                curleft = curright\n",
    "                curright = seq_len\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            if len(canabd)==1:\n",
    "                abdlist = abdlist + canabd\n",
    "            else:\n",
    "                curabd = [canabd[0][0],canabd[0][1],canabd[0][2],canabd[0][3],canabd[0][4],canabd[0][5]]\n",
    "                for k in range(1,len(canabd)):\n",
    "                    if canabd[k][4] != canabd[k-1][4]:\n",
    "                        abdlist.append(curabd)\n",
    "                        curabd = [canabd[k][0],canabd[k][1],canabd[k][2],canabd[k][3],canabd[k][4],canabd[k][5]]\n",
    "                    else:\n",
    "                        curabd[3] = canabd[k][3]\n",
    "                        curabd[5] = curabd[5] + canabd[k][5]\n",
    "                    if k == len(canabd)-1:\n",
    "                        abdlist.append(curabd)\n",
    "\n",
    "\n",
    "\n",
    "#             print(\"abdlist sample:\",i)\n",
    "    print(\"numibd\",len(abdlist))\n",
    "    \n",
    "    abduse = abdlist\n",
    "\n",
    "#########################################################################\n",
    "\n",
    "    tolnummu = 0\n",
    "    for i in range(len(abduse)):\n",
    "        tolnummu = tolnummu+len(abduse[i][5])\n",
    "\n",
    "\n",
    "    muloclist = np.zeros(tolnummu)\n",
    "    indd = 0\n",
    "    for i in range(len(abduse)):\n",
    "        abdc = abduse[i]\n",
    "        lee = len(abdc[5])\n",
    "        muloclist[indd:(indd+lee)] = abdc[5]\n",
    "        indd = indd+lee\n",
    "\n",
    "    smulist = numpy.sort(muloclist)\n",
    "    count2 = 0\n",
    "    curloc = -1\n",
    "    curcount = 1\n",
    "    for i in range(len(smulist)):\n",
    "        if smulist[i] == curloc:\n",
    "            curcount = curcount+1\n",
    "        else:\n",
    "            if curcount ==2:\n",
    "                count2 = count2+1\n",
    "            curcount = 1\n",
    "            curloc = smulist[i]\n",
    "\n",
    "#         summary_sta1_sim = (len(smulist) - 2*count2)/len(abduse)        \n",
    "#         summary_sta2_sim = (2*count2) / len(abduse)\n",
    "\n",
    "    summary_sta1_inf = (len(smulist) - 2*count2)        \n",
    "    summary_sta2_inf = (2*count2)  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "########################################################################            \n",
    "\n",
    "\n",
    "\n",
    "    lenlist = numpy.zeros(len(abduse))\n",
    "    for i in range(len(abduse)):\n",
    "        lenlist[i] = abduse[i][3] - abduse[i][2]\n",
    "\n",
    "    summary_sta3_inf = statistics.mean(lenlist) \n",
    "    summary_sta4_inf = statistics.stdev(lenlist) \n",
    "\n",
    "#################################################################################################\n",
    "    stime = time.time()\n",
    "\n",
    "    for run in range(simtime):\n",
    "        \n",
    "        popsim = numpy.random.uniform(priorinipop[0],priorinipop[1])\n",
    "\n",
    "#         if run == 0:\n",
    "#             popsim = 109000\n",
    "\n",
    "#         if run == 1:\n",
    "#             popsim = 109000\n",
    "\n",
    "#         if run == 2:\n",
    "#             popsim = 4000\n",
    "\n",
    "#         if run == 3:\n",
    "#             popsim = 4000\n",
    "           \n",
    "#         if run == 4:\n",
    "#             popsim = 50000\n",
    "         \n",
    "#         if run == 5:\n",
    "#             popsim = 50000\n",
    "   \n",
    "        \n",
    "        \n",
    "        \n",
    "        inipopsimlist.append(popsim)\n",
    "\n",
    "        \n",
    "        \n",
    "#         demo_model = msprime.Demography.isolated_model([popsim], growth_rate=[ratesim])\n",
    "#         demo_model.add_population_parameters_change(800,  initial_size=None, growth_rate=0, population=None)\n",
    "\n",
    "\n",
    "        genpoint = [0*5/5.2,400*5/5.2,800*5/5.2,1200*5/5.2,1600*5/5.2,2000*5/5.2,2400*5/5.2,2800*5/5.2,\\\n",
    "                    3200*5/5.2,3600*5/5.2,4000*5/5.2,8000*5/5.2,12000*5/5.2,16000*5/5.2,20000*5/5.2]\n",
    "        inipopsize =[15200*5/5.2, 15200*5/5.2, 15600*5/5.2, 14800*5/5.2, 12400*5/5.2, 10400*5/5.2,9800*5/5.2,\\\n",
    "                     10400*5/5.2,12000*5/5.2,14400*5/5.2,16800*5/5.2,29400*5/5.2,28000*5/5.2,22000*5/5.2,18000*5/5.2]\n",
    "        genuse = []\n",
    "        inipuse = []\n",
    "        gruse = []\n",
    "        for i in range(len(genpoint)-1):\n",
    "            addplist = []\n",
    "            for j in range(10):\n",
    "                addplist.append(genpoint[i] + (genpoint[i+1]-genpoint[i])/10*j)\n",
    "            genuse = genuse + addplist               \n",
    "        for i in range(len(inipopsize)-1):\n",
    "            addplist = []\n",
    "            for j in range(10):\n",
    "                addplist.append(inipopsize[i] + (inipopsize[i+1]-inipopsize[i])/10*j)\n",
    "            inipuse = inipuse + addplist\n",
    "        genuse.append(genpoint[len(genpoint)-1])\n",
    "        inipuse.append(inipopsize[len(inipopsize)-1])\n",
    "\n",
    "        \n",
    "        ind1000 = 0\n",
    "        for i in range(len(genuse)):\n",
    "            if genuse[i]<1000:\n",
    "                ind1000 = i\n",
    "        ratesim =   -numpy.log(inipuse[ind1000]/popsim )/1000   \n",
    "        for i in range(len(genuse)):\n",
    "            if genuse[i]<1000:\n",
    "                inipuse[i] = popsim*numpy.exp(-ratesim*genuse[i]) \n",
    "        for i in range(len(genuse)-1):\n",
    "            gruse.append( numpy.log( inipuse[i]/inipuse[i+1] )/(genuse[i+1] - genuse[i])  )\n",
    "        gruse.append(0)\n",
    "        \n",
    "\n",
    "        demo_model = msprime.Demography.isolated_model([inipuse[0]], growth_rate=[gruse[0]])\n",
    "        for i in range(1,len(genuse)):\n",
    "            demo_model.add_population_parameters_change(genuse[i],  initial_size=inipuse[i], growth_rate=gruse[i], population=None)\n",
    "\n",
    "                \n",
    "                \n",
    "\n",
    "#         for i in range(len(genuse)-1):\n",
    "#             gruse.append( numpy.log( inipuse[i]/inipuse[i+1] )/(genuse[i+1] - genuse[i])  )\n",
    "#         gruse.append(0)\n",
    "        \n",
    "\n",
    "#         demo_model = msprime.Demography.isolated_model([inipuse[0]], growth_rate=[gruse[0]])\n",
    "#         for i in range(1,len(genuse)):\n",
    "#             demo_model.add_population_parameters_change(genuse[i],  initial_size=inipuse[i], growth_rate=gruse[i], population=None)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ts = msprime.sim_ancestry(\n",
    "            samples=sam_size,\n",
    "            recombination_rate= r, \n",
    "            sequence_length= seq_len,\n",
    "            record_provenance=False,\n",
    "#             population_size = pop_size,\n",
    "#             gene_conversion_rate = gene_conver ,\n",
    "#             gene_conversion_tract_length = track_len,\n",
    "            #random_seed =seed,\n",
    "            #discrete_genome=False\n",
    "            demography = demo_model\n",
    "            )\n",
    "        # Visualise the simulated ancestral history.\n",
    "        #SVG(ts.draw_svg())\n",
    "\n",
    "        #ts.num_trees\n",
    "        \n",
    "        print(\"done gene\")\n",
    "\n",
    "        mts = msprime.sim_mutations(ts, rate = mu,\n",
    "                                    #discrete_genome=False,\n",
    "                                    #random_seed=seed\n",
    "                                  )\n",
    "\n",
    "        print(\"done mu\")\n",
    "\n",
    "        sample_data = tsinfer.SampleData.from_tree_sequence(mts, use_sites_time=None, use_individuals_time=None)\n",
    "\n",
    "    \n",
    "        sitefreq = numpy.sum(sample_data.sites_genotypes,1)\n",
    "        siteuse =  numpy.where( (sitefreq>freqconsider[0] ) * ( sitefreq<freqconsider[1] ) )[0]  \n",
    "        print(\"siteuse_sim\",len(siteuse))\n",
    "\n",
    "        datause = sample_data.subset(sites = siteuse)\n",
    "\n",
    "        inferred_ts = tsinfer.infer(datause,path_compression=False)\n",
    "\n",
    "\n",
    "\n",
    "        mts = inferred_ts\n",
    "\n",
    "        ###############################################################################\n",
    "        msites = mts.tables.mutations.site\n",
    "        sitepos = mts.tables.sites.position\n",
    "        mpos = []\n",
    "        for i in range(len(msites)):\n",
    "            cursite = msites[i]\n",
    "            mpos.append(sitepos[cursite])\n",
    "        mpos = numpy.array(mpos)\n",
    "        mnodes = mts.tables.mutations.node\n",
    "        ec = mts.tables.edges.child\n",
    "        ep = mts.tables.edges.parent\n",
    "        el = mts.tables.edges.left\n",
    "        er = mts.tables.edges.right\n",
    "\n",
    "\n",
    "        leftmost = min(el)\n",
    "        rightmost = max(er)\n",
    "        for i in range(len(el)):\n",
    "            if el[i] == leftmost:\n",
    "                el[i] = 0\n",
    "            if er[i] == rightmost:\n",
    "                er[i] = seq_len\n",
    "\n",
    "\n",
    "        if len(mpos) == 0:\n",
    "            mpos_sn = []\n",
    "            mnodes_sn = []\n",
    "        else:\n",
    "            mnp_sn = numpy.transpose(sorted(numpy.transpose(numpy.array([mnodes,mpos])).tolist()))\n",
    "            mnodes_sn = mnp_sn[0]\n",
    "            mpos_sn = mnp_sn[1]\n",
    "\n",
    "        eclrp_sc = numpy.transpose(sorted(numpy.transpose(numpy.array([ec,el,er,ep])).tolist()))\n",
    "        ec_sc = eclrp_sc[0]\n",
    "        el_sc = eclrp_sc[1]\n",
    "        er_sc = eclrp_sc[2]\n",
    "        ep_sc = eclrp_sc[3]\n",
    "        enm_sc = [[] for i in range(len(ec_sc))]\n",
    "\n",
    "        curedge = 0\n",
    "        for i in range(len(mnodes_sn)):\n",
    "            curnodemu = mnodes_sn[i] \n",
    "            cursitemu = mpos_sn[i]\n",
    "            for j in range(curedge,len(ec_sc)):\n",
    "                if curnodemu == ec_sc[j]:\n",
    "\n",
    "                    if er_sc[j] > cursitemu and el_sc[j] <= cursitemu:\n",
    "\n",
    "                        enm_sc[j].append(cursitemu)  \n",
    "\n",
    "                        curedge = j\n",
    "                        break\n",
    "                    if el_sc[j] > cursitemu:\n",
    "                        curedge = j\n",
    "                        break\n",
    "\n",
    "                if ec_sc[j] > mnodes_sn[i] or j == len(ec_sc)-1:\n",
    "                    #print(\"missing mutations\")\n",
    "                    curedge = j\n",
    "                    break\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ###########################################################################\n",
    "        ###########################################################################\n",
    "        ##########################################################################\n",
    "\n",
    "        clocate = []\n",
    "        curc = -1\n",
    "        for i in range(len(ec_sc)):\n",
    "            while ec_sc[i] != curc:\n",
    "                curc=curc+1\n",
    "                clocate.append(i)\n",
    "        clocate.append(len(ec_sc))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        abdlist = []\n",
    "\n",
    "        for i in range(0,2*sam_size-1):\n",
    "            for j in range(i+1,i+2):\n",
    "                canabd = []\n",
    "                curleft = 0\n",
    "                curright =seq_len\n",
    "                #start = 0\n",
    "                while curleft != seq_len:\n",
    "                    curp1 = i\n",
    "                    curp2 = j \n",
    "                    curmed = []\n",
    "                    while curp1 != curp2:\n",
    "                        if curp1 < curp2:\n",
    "                            begin1 = clocate[curp1]\n",
    "                            end = clocate[curp1+1]\n",
    "                            s_range = el_sc[begin1:end]\n",
    "                            begin2 = find_s_se(curleft,s_range)\n",
    "                            for k in range(begin1+begin2,end):\n",
    "                                if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                    curp1 = int(ep_sc[k])\n",
    "                                    curoldr = curright\n",
    "                                    curright = min(er_sc[k],curright)\n",
    "\n",
    "                                    curmed.append(k)\n",
    "\n",
    "                                    break\n",
    "\n",
    "                        else:\n",
    "                            begin1 = clocate[curp2]\n",
    "                            end = clocate[curp2+1]\n",
    "                            s_range = el_sc[begin1:end]\n",
    "                            begin2 = find_s_se(curleft,s_range)\n",
    "                            for k in range(begin1+begin2,end):\n",
    "                                if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                    curp2 = int(ep_sc[k])\n",
    "                                    curoldr = curright\n",
    "                                    curright = min(er_sc[k],curright) \n",
    "\n",
    "                                    curmed.append(k)\n",
    "\n",
    "                                    break\n",
    "\n",
    "                    canmution = []\n",
    "                    for k in range(len(curmed)):\n",
    "                        edgemu = enm_sc[curmed[k]]\n",
    "                        startt = find_s_s(curleft,edgemu)\n",
    "                        endd = find_s_s(curright,edgemu)                    \n",
    "                        canmution = canmution + edgemu[startt:endd] \n",
    "\n",
    "                    canabd.append([i,j,curleft,curright,curp1,canmution])\n",
    "                    curleft = curright\n",
    "                    curright = seq_len\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "                if len(canabd)==1:\n",
    "                    abdlist = abdlist + canabd\n",
    "                else:\n",
    "                    curabd = [canabd[0][0],canabd[0][1],canabd[0][2],canabd[0][3],canabd[0][4],canabd[0][5]]\n",
    "                    for k in range(1,len(canabd)):\n",
    "                        if canabd[k][4] != canabd[k-1][4]:\n",
    "                            abdlist.append(curabd)\n",
    "                            curabd = [canabd[k][0],canabd[k][1],canabd[k][2],canabd[k][3],canabd[k][4],canabd[k][5]]\n",
    "                        else:\n",
    "                            curabd[3] = canabd[k][3]\n",
    "                            curabd[5] = curabd[5] + canabd[k][5]\n",
    "                        if k == len(canabd)-1:\n",
    "                            abdlist.append(curabd)\n",
    "\n",
    "\n",
    "    #             print(\"abdlist sample:\",i)\n",
    "        print(\"numibd\",len(abdlist))\n",
    "\n",
    "        abduse = abdlist\n",
    "\n",
    "\n",
    "        tolnummu = 0\n",
    "        for i in range(len(abduse)):\n",
    "            tolnummu = tolnummu+len(abduse[i][5])\n",
    "\n",
    "\n",
    "        muloclist = np.zeros(tolnummu)\n",
    "        indd = 0\n",
    "        for i in range(len(abduse)):\n",
    "            abdc = abduse[i]\n",
    "            lee = len(abdc[5])\n",
    "            muloclist[indd:(indd+lee)] = abdc[5]\n",
    "            indd = indd+lee\n",
    "            \n",
    "        smulist = numpy.sort(muloclist)\n",
    "        count2 = 0\n",
    "        curloc = -1\n",
    "        curcount = 1\n",
    "        for i in range(len(smulist)):\n",
    "            if smulist[i] == curloc:\n",
    "                curcount = curcount+1\n",
    "            else:\n",
    "                if curcount ==2:\n",
    "                    count2 = count2+1\n",
    "                curcount = 1\n",
    "                curloc = smulist[i]\n",
    "        \n",
    "#         summary_sta1_sim = (len(smulist) - 2*count2)/len(abduse)        \n",
    "#         summary_sta2_sim = (2*count2) / len(abduse)\n",
    "\n",
    "        summary_sta1_sim = (len(smulist) - 2*count2)        \n",
    "        summary_sta2_sim = (2*count2)  \n",
    "\n",
    "        summary_sta1_simlist.append(summary_sta1_sim)\n",
    "        summary_sta2_simlist.append(summary_sta2_sim)\n",
    "        \n",
    "        \n",
    "########################################################################            \n",
    "        \n",
    "        \n",
    "\n",
    "        lenlist = numpy.zeros(len(abduse))\n",
    "        for i in range(len(abduse)):\n",
    "            lenlist[i] = abduse[i][3] - abduse[i][2]\n",
    "\n",
    "        summary_sta3_simlist.append( statistics.mean(lenlist) )\n",
    "        summary_sta4_simlist.append( statistics.stdev(lenlist) )\n",
    "        \n",
    "        print(\"-----------------------------------\")\n",
    "\n",
    "#         print(inipopsimlist )\n",
    "#         print(iniratesimlist )\n",
    "        print(summary_sta1_simlist,summary_sta1_inf)\n",
    "        print(summary_sta2_simlist,summary_sta2_inf)\n",
    "        print(summary_sta3_simlist,summary_sta3_inf)\n",
    "        print(summary_sta4_simlist,summary_sta4_inf)\n",
    "        print(time.time()-stime, run)\n",
    "        print(\"-----------------------------------\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "byrun = 1\n",
    "name1 = \"1M2_ne_realpopsimall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name3 = \"1M2_ne_realsta1simall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name4 = \"1M2_ne_realsta2simall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name5 = \"1M2_ne_realsta1inf_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name6 = \"1M2_ne_realsta2inf_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name7 = \"1M2_ne_realsta3simall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name8 = \"1M2_ne_realsta4simall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name9 = \"1M2_ne_realsta3inf_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "name10 = \"1M2_ne_realsta4inf_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "file = open(name1,\"wb\") \n",
    "pickle.dump(inipopsimlist,file)\n",
    "file.close()\n",
    "file = open(name3,\"wb\") \n",
    "pickle.dump(summary_sta1_simlist,file)\n",
    "file.close()\n",
    "file = open(name4,\"wb\") \n",
    "pickle.dump(summary_sta2_simlist,file)\n",
    "file.close()\n",
    "file = open(name5,\"wb\") \n",
    "pickle.dump(summary_sta1_inf,file)\n",
    "file.close()\n",
    "file = open(name6,\"wb\") \n",
    "pickle.dump(summary_sta2_inf,file)\n",
    "file.close()\n",
    "file = open(name7,\"wb\") \n",
    "pickle.dump(summary_sta3_simlist,file)\n",
    "file.close()\n",
    "file = open(name8,\"wb\") \n",
    "pickle.dump(summary_sta4_simlist,file)\n",
    "file.close()\n",
    "file = open(name9,\"wb\") \n",
    "pickle.dump(summary_sta3_inf,file)\n",
    "file.close()\n",
    "file = open(name10,\"wb\") \n",
    "pickle.dump(summary_sta4_inf,file)\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "856de915",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54365.636569180904"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.exp(0.001*1000)*20000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e37f6a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f027283",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "7f0ae5f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0011417817431495782"
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.log(17400/2800)/1600"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "283ac458",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7acf9e29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[131917.87691659856, 137940.00308222283] 83605.94053328519\n"
     ]
    }
   ],
   "source": [
    "print(summary_sta1_simlist,summary_sta1_inf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ae8eb346",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[319061.2179425511, 326424.3238242735] 240396.64324638573\n"
     ]
    }
   ],
   "source": [
    "print(summary_sta2_simlist,summary_sta2_inf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04bc6fe3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22143835",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e0c6ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d2b979c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14825b48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6210252a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0310ab23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "252e8a30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aebcc7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6199afb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "f424e133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[76000, 76000, 30000, 30000, 120000, 120000]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inipopsimlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "5c84dcd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.0033, 0.0033, 0.004, 0.004, 0.002, 0.002]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iniratesimlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0388516",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "5c650ad9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10886.154394729501"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "120000*numpy.exp(-1200*0.002) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "ae18e951",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2344345.562264151,\n",
       " 2378972.1105032824,\n",
       " 2438101.860213044,\n",
       " 2414636.8784008883,\n",
       " 1396668.7778610426,\n",
       " 1251081.9959723821]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_sta1_simlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "e603cf37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4372968.630850762,\n",
       " 4517109.553498169,\n",
       " 4640203.3000262,\n",
       " 4673087.035534496,\n",
       " 2726643.539254386,\n",
       " 2469605.7512018387]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_sta2_simlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1fcc003",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d5fa72b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "363acbaa",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "path                  = None\n",
      "file_size             = -1 Byte\n",
      "format_name           = tsinfer-sample-data\n",
      "format_version        = (5, 0)\n",
      "finalised             = True\n",
      "uuid                  = b9ca7636-ecf7-46e1-8ee7-49341e38a07c\n",
      "num_provenances       = 2\n",
      "provenances/timestamp = shape=(2,); dtype=object;\n",
      "provenances/record    = shape=(2,); dtype=object;\n",
      "sequence_length       = 63025522.0\n",
      "metadata_schema       = {'additionalProperties': True, 'codec': 'json', 'properties': {}, 'required': [], 'type': 'object'}\n",
      "metadata              = {}\n",
      "num_populations       = 26\n",
      "num_individuals       = 104\n",
      "num_samples           = 208\n",
      "num_sites             = 1552394\n",
      "populations/metadata_schema = None\n",
      "populations/metadata  = shape=(26,); dtype=object;\n",
      "individuals/metadata_schema = None\n",
      "individuals/metadata  = shape=(104,); dtype=object;\n",
      "individuals/location  = shape=(104,); dtype=object;\n",
      "individuals/time      = shape=(104,); dtype=float64;uncompressed size=832 Bytes\n",
      "individuals/population = shape=(104,); dtype=int32;uncompressed size=416 Bytes\n",
      "individuals/flags     = shape=(104,); dtype=uint32;uncompressed size=416 Bytes\n",
      "samples/individual    = shape=(208,); dtype=int32;uncompressed size=832 Bytes\n",
      "sites/position        = shape=(1552394,); dtype=float64;uncompressed size=12.4 MB\n",
      "sites/time            = shape=(1552394,); dtype=float64;uncompressed size=12.4 MB\n",
      "sites/alleles         = shape=(1552394,); dtype=object;\n",
      "sites/genotypes       = shape=(1552394, 208); dtype=int8;uncompressed size=322.9 MB\n",
      "sites/metadata_schema = None\n",
      "sites/metadata        = shape=(1552394,); dtype=object;\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(newdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5dabf35c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "208"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(numpy.sum(newdata.sites_genotypes,1) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "04406907",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa = numpy.sum(newdata.sites_genotypes,1)<2\n",
    "sitepos = newdata.sites_position\n",
    "sitegen = newdata.sites_genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ab8cd837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1375936"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "4e579ee8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1552394"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "5bcee209",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "176000"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1552000-1376000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "358aeea1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.90507049112532"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(aa)/len(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1d28549a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan, nan, nan])"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newdata.sites_time[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09496002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, ...,  True, False,  True])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "72a8febd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "17%10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3929d48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "627e8f17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "590e03a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "541cf616",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 2, 3],\n",
       "       [4, 5, 6]])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa = numpy.array( [ [1,2,3],[4,5,6]  ])\n",
    "aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b945089f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 6, 15])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.sum(aa,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "61dcd48f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>237090</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>63025522.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>208</td></tr>\n",
       "                      <tr><td>Total Size</td><td>51.4 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>836673</td>\n",
       "                      <td>25.5 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>104</td>\n",
       "                      <td>2.9 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>217501</td>\n",
       "                      <td>7.7 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>143513</td>\n",
       "                      <td>3.8 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>224 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>2</td>\n",
       "                      <td>2.8 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>217095</td>\n",
       "                      <td>5.2 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7f45d9f925b0>"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtsjjj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "63b14fd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_datajjj = tsinfer.SampleData.from_tree_sequence(mtsjjj, use_sites_time=True, use_individuals_time=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "49c39fca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "116583"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(sample_datajjj.sites_time[0:len(sample_datajjj.sites_time)]<1200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "3a14e5b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "151927"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(sample_datajjj.sites_time[0:len(sample_datajjj.sites_time)]<2400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "ca739ab9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "217095"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sample_datajjj.sites_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e691f161",
   "metadata": {},
   "outputs": [],
   "source": [
    "genty = sample_datajjj.sites_genotypes\n",
    "sitetime = sample_datajjj.sites_time\n",
    "\n",
    "sitefreq = numpy.sum(genty,1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b29635b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec3ab03d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "419dcd26",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "b7175606",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8350550134240602"
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( (sitefreq < 100)*(sitetime[:]<2400) )/ sum((sitefreq < 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "bea3ae6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9932123891976775"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( (sitefreq < 6)*(sitetime[:]<1200) )/ sum((sitefreq < 6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "d6eb0f6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30663"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " sum((sitefreq < 6)*(sitefreq >1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f5f8fea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "7bf0f9ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "sitefreq1 = numpy.sum(newdata.sites_genotypes,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "5930b43f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33408"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " sum((sitefreq1 < 6)*(sitefreq1 >1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae4f7034",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "837a5924",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "209618"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sample_data.sites_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f27ba50",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b69f8d23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "200b61eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb3bfab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e19e407a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "c97deb8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.30985387785687524"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( (sitefreq < 2) )/len(sitefreq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "3e270dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "202844"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sitefreq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "2f1ba789",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "141990.8"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "202844*0.7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c29361be",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa = [True, False, True, True, False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e902bd8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 2, 3])"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.where(aa)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bcb4027",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a31312",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa398c5f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb317db8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0b25ddf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "3c50b127",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_data = tsinfer.SampleData.from_tree_sequence(mts, use_sites_time=True, use_individuals_time=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "a5d6e902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "110997"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sample_data.sites_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "f0fc897f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sitefreq2 = numpy.sum(sample_data.sites_genotypes,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "0b9aab6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23786"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum((sitefreq2 < 4)*(sitefreq2 >1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "158c5210",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9987745098039216"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sitetime2 = sample_data.sites_time\n",
    "\n",
    "sum( (sitefreq2 < 200)*(sitefreq2 >1)*(sitetime2[:]<1200) )/ sum((sitefreq2 < 200)*(sitefreq2 >1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "09372f4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "110026"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(sitetime2[:]<1200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "fa1f752c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "110161"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sitetime2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "e0b7eb10",
   "metadata": {},
   "outputs": [],
   "source": [
    "sitefreq1 = numpy.sum(newdata.sites_genotypes,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "8eb62a56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23701"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum((sitefreq1 < 4)*(sitefreq1 >1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d76369b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "4433aa0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0031499974619990877"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.log(17400/1400)/800"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "a5b33a5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.3300000000000002e-08"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "5c8a7ea3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "63025522"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07d2898b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "904170f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f009322a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a20de92b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49fda1e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4dd19c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "048991a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "930e96ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0b19c46",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2bdb18d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import statistics\n",
    "import msprime\n",
    "from IPython.display import SVG, display\n",
    "import tsinfer\n",
    "import scipy\n",
    "import math\n",
    "import numpy\n",
    "import tskit\n",
    "import io\n",
    "import builtins\n",
    "import sys\n",
    "from tqdm.notebook import tqdm\n",
    "from tskit import MISSING_DATA\n",
    "import pickle\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import cvxopt as opt\n",
    "from cvxopt import matrix, spmatrix, sparse\n",
    "from cvxopt.solvers import qp, options\n",
    "from cvxopt import blas\n",
    "from cvxopt import spdiag\n",
    "import statsmodels\n",
    "from statsmodels.nonparametric.kernel_regression import KernelReg\n",
    "from scipy.signal import savgol_filter\n",
    "import time\n",
    "import bisect\n",
    "\n",
    "\n",
    "def find_s(elem, sorted_list):\n",
    "    'Locate the leftmost value exactly equal to x'\n",
    "    i = bisect.bisect_left(sorted_list, elem)\n",
    "    if i != len(sorted_list) and sorted_list[i] == elem:\n",
    "        return i\n",
    "\n",
    "\n",
    "def find_s_se(elem, sorted_list):\n",
    "    'Locate the leftmost value smaller or equal to x'\n",
    "    i = bisect.bisect_right(sorted_list, elem)\n",
    "    if i != 0:\n",
    "        return i-1\n",
    "    else:\n",
    "        print(\"error\")\n",
    "\n",
    "\n",
    "def find_s_s(elem, sorted_list):\n",
    "    'Locate the leftmost value smaller or equal to x'\n",
    "    i = bisect.bisect_left(sorted_list, elem)\n",
    "    if i != 0:\n",
    "        return i\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    " \n",
    "\n",
    "    \n",
    "file = open(\"recombination_map.dat\",\"rb\")\n",
    "rmap = pickle.load(file)\n",
    "file.close\n",
    "\n",
    "sam=tsinfer.load(\"tgp_chr20.samples\")\n",
    "\n",
    "LWKlist = []\n",
    "for i in range(len(sam.individuals_population)):\n",
    "    if sam.individuals_population[i]==11:\n",
    "        LWKlist.append(i)\n",
    "newdata = sam.subset(individuals=LWKlist)\n",
    "\n",
    "        \n",
    "    \n",
    "    \n",
    "seq_len = 63025522\n",
    "sam_size = 99\n",
    "r = rmap\n",
    "mu = 1.375*10**(-8)\n",
    "# gene_conver = 10**(-8)\n",
    "# track_len = 100\n",
    "# epsilon = 0.0001\n",
    "totaliter=1\n",
    "\n",
    "simtime = 2\n",
    "# priormu = [1.0*10**(-8), 2.0*10**(-8)]\n",
    "# prioreps = [0, 0.0002]\n",
    "# accept_threshold = [1000,400]\n",
    "freqconsider = [1,10]\n",
    "priorinipop = [120000, 5000]\n",
    "priorinirate = [-0.001 , 0.004]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "663f9acc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5463dd4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The provenance information for the resulting tree sequence is 2.78MB. This is nothing to worry about as provenance is a good thing to have, but if you want to save this memory/storage space you can disable provenance recording by setting record_provenance=False\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "genpoint = [0*5/5.2,400*5/5.2,800*5/5.2,1200*5/5.2,1600*5/5.2,2000*5/5.2,2400*5/5.2,2800*5/5.2,\\\n",
    "            3200*5/5.2,3600*5/5.2,4000*5/5.2,8000*5/5.2,12000*5/5.2,16000*5/5.2,20000*5/5.2]\n",
    "inipopsize =[15200*5/5.2, 15200*5/5.2, 15600*5/5.2, 14800*5/5.2, 12400*5/5.2, 10400*5/5.2,9800*5/5.2,\\\n",
    "             10400*5/5.2,12000*5/5.2,14400*5/5.2,16800*5/5.2,29400*5/5.2,28000*5/5.2,22000*5/5.2,18000*5/5.2]\n",
    "genuse = []\n",
    "inipuse = []\n",
    "gruse = []\n",
    "for i in range(len(genpoint)-1):\n",
    "    addplist = []\n",
    "    for j in range(10):\n",
    "        addplist.append(genpoint[i] + (genpoint[i+1]-genpoint[i])/10*j)\n",
    "    genuse = genuse + addplist\n",
    "for i in range(len(inipopsize)-1):\n",
    "    addplist = []\n",
    "    for j in range(10):\n",
    "        addplist.append(inipopsize[i] + (inipopsize[i+1]-inipopsize[i])/10*j)\n",
    "    inipuse = inipuse + addplist\n",
    "genuse.append(genpoint[len(genpoint)-1])\n",
    "inipuse.append(inipopsize[len(inipopsize)-1])\n",
    "for i in range(len(genuse)-1):\n",
    "    gruse.append( numpy.log( inipuse[i]/inipuse[i+1] )/(genuse[i+1] - genuse[i])  )\n",
    "gruse.append(0.0000005)\n",
    "    \n",
    "\n",
    "demo_model = msprime.Demography.isolated_model([inipuse[0]], growth_rate=[gruse[0]])\n",
    "for i in range(1,len(genuse)):\n",
    "    demo_model.add_population_parameters_change(genuse[i],  initial_size=inipuse[i], growth_rate=gruse[i], population=None)\n",
    "\n",
    "    \n",
    "\n",
    "ts = msprime.sim_ancestry(\n",
    "        samples=sam_size,\n",
    "        recombination_rate= r, \n",
    "        sequence_length= seq_len,\n",
    "#         population_size = pop_size,\n",
    "#         gene_conversion_rate = gene_conver ,\n",
    "#         gene_conversion_tract_length = track_len,\n",
    "        #random_seed =seed,\n",
    "        #discrete_genome=False\n",
    "        demography = demo_model\n",
    "        )\n",
    "    # Visualise the simulated ancestral history.\n",
    "    #SVG(ts.draw_svg())\n",
    "\n",
    "    #ts.num_trees\n",
    "\n",
    "mtsjjj = msprime.sim_mutations(ts, rate = mu,\n",
    "                            #discrete_genome=False,\n",
    "                            #random_seed=seed\n",
    "                          )\n",
    "sample_datajjj = tsinfer.SampleData.from_tree_sequence(mtsjjj, use_sites_time=True, use_individuals_time=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8efa40db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "109035"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(sample_datajjj.sites_time[0:len(sample_datajjj.sites_time)]<1000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "16c83263",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "151123"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(sample_datajjj.sites_time[0:len(sample_datajjj.sites_time)]<2400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ca944909",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "301949"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sample_datajjj.sites_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b53ae11",
   "metadata": {},
   "outputs": [],
   "source": [
    "genty = sample_datajjj.sites_genotypes\n",
    "sitetime = sample_datajjj.sites_time\n",
    "\n",
    "sitefreq = numpy.sum(genty,1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a961d477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.13127978014591035"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum((sitefreq>6)*(sitefreq < 40)*(sitetime[:]<1000) )/ sum((sitefreq < 40)*(sitefreq>6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6f19b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum( (sitefreq>1) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f18fe3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(sample_datajjj.sites_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7dc2d0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc0adee0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ecce5dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genty1 = newdata.sites_genotypes\n",
    "sitetime1 = newdata.sites_time\n",
    "\n",
    "sitefreq1 = numpy.sum(genty1,1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c85d6454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "235622"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( (sitefreq1>1)*(sitefreq1<40) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9b432727",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "181490"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( (sitefreq>1)*(sitefreq<40) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8d98b746",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.1233843907288883"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "144629/128744"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa5da94e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7aa2f23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "821171dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.375e-08"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "64460f8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7ff181af83a0>]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD4CAYAAADsKpHdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAq/0lEQVR4nO3deXxV5bX/8c/KAAlDAgkhhARIEAQBESQiONWKVrRacKjFaqEWS7Vea7X3evXXV+/tbe9t1dtqq1ZbqrVo68ClDtSKiqB1QiCoyCxhDkMSphACBBLW74/zxIY0kgCBk5x836/Xee191tnPznoiZp39PHswd0dERCQu2gmIiEjzoIIgIiKACoKIiAQqCCIiAqggiIhIkBDtBI5Wly5dPDc3N9ppiIi0KAsWLNjq7hn1fdZiC0Jubi4FBQXRTkNEpEUxs3Wf95mGjEREBFBBEBGRQAVBREQAFQQREQlUEEREBGhkQTCz281siZktNrNnzCzJzNLMbKaZrQzLzrW2v9vMCs1shZldXCs+zMwWhc8eNDML8bZm9lyIzzWz3CbvqYiIHFaDBcHMsoHvAfnuPgiIB8YBdwGz3L0vMCu8x8wGhM8HAqOBR8wsPuzuUWAS0De8Rof4RGCHu/cBHgDubZLeiYhIozV2yCgBSDazBKAdsAkYA0wJn08Bxob1McCz7l7p7muAQmC4mWUBKe4+xyP33H6yTpuafU0DRtUcPTS1xRvLuPfV5ei23yIih2qwILj7RuAXwHpgM1Dm7q8Dme6+OWyzGegammQDG2rtoijEssN63fghbdy9CigD0uvmYmaTzKzAzApKS0sb28dDfLh+B4++tYo5q7YdVXsRkVjVmCGjzkS+wecB3YH2Znb94ZrUE/PDxA/X5tCA+2R3z3f3/IyMeq+8btA1+T3ITGnLr2atPKr2IiKxqjFDRhcCa9y91N0PAM8DZwHFYRiIsCwJ2xcBPWq1zyEyxFQU1uvGD2kThqVSge1H06GGJCXGc9MXTmLemu18sFpHCSIiNRpTENYDI8ysXRjXHwUsA6YDE8I2E4CXwvp0YFw4cyiPyOTxvDCsVG5mI8J+xtdpU7Ovq4HZfhwH+a8d3pOMjm359Rs6ShARqdGYOYS5RCZ6PwQWhTaTgXuAi8xsJXBReI+7LwGmAkuBV4Fb3L067O5m4DEiE82rgBkh/jiQbmaFwB2EM5aOl5qjhDmrtzFvzXE5EBERaXGspZ5tk5+f78dyt9O9+6s597436detA3++cUQTZiYi0nyZ2QJ3z6/vs1Z7pXJym3hu+kJv3ivcxvy1OkoQEWm1BQHgujN7kdGxLb98fUW0UxERibpWXRCS28Tz3fNP4oPV23l/1dZopyMiElWtuiBA5IyjbilJ3P/6p7p6WURatVZfEJIS47nlgj4UrNvBOyt1lCAirVerLwgAX8vvQXanZH45U0cJItJ6qSAAbRLiuPWCPizcsJM3lpU03EBEJAapIARXDcshr0t7fvn6Cg4e1FGCiLQ+KghBYnwct190Msu3lDN94aaGG4iIxBgVhFouOzWLU7JSuH/mpxyoPhjtdERETigVhFri4ox/u/hk1m/fw3PzNzTcQEQkhqgg1PHFfl3J79WZB2etZO/+6oYbiIjECBWEOsyMO0f3p6S8kifeXxPtdEREThgVhHoMz0tjVP+uPPrWKnZU7I92OiIiJ4QKwue4c3R/Kiqr+M2bhdFORUTkhFBB+Bz9unXkqtNzeHLOOjZs3xPtdEREjjsVhMO440snYwYPzPw02qmIiBx3DRYEM+tnZh/Xeu0ys++bWZqZzTSzlWHZuVabu82s0MxWmNnFteLDzGxR+OzB8GxlwvOXnwvxuWaWe1x6e4SyUpO54ew8Xvh4I0s2lUU7HRGR46oxz1Re4e5D3H0IMAzYA7xA5LnHs9y9LzArvMfMBgDjgIHAaOARM4sPu3sUmAT0Da/RIT4R2OHufYAHgHubpHdN4ObzTyI1OZGfvbJMN74TkZh2pENGo4BV7r4OGANMCfEpwNiwPgZ41t0r3X0NUAgMN7MsIMXd53jkL+uTddrU7GsaMKrm6CHaUpMTuW1UX94r3MabK3TjOxGJXUdaEMYBz4T1THffDBCWXUM8G6h9mW9RiGWH9brxQ9q4exVQBqTX/eFmNsnMCsysoLS09AhTP3rXndmLvC7t+dkry6nSLS1EJEY1uiCYWRvgK8D/NbRpPTE/TPxwbQ4NuE9293x3z8/IyGggjabTJiGOuy7pT2HJbp7RLS1EJEYdyRHCJcCH7l4c3heHYSDCsmY8pQjoUatdDrApxHPqiR/SxswSgFRg+xHkdtx9aUAmw/PS+NXMTynfdyDa6YiINLkjKQjX8o/hIoDpwISwPgF4qVZ8XDhzKI/I5PG8MKxUbmYjwvzA+DptavZ1NTDbm9kMrpnxoy8PYFvFfh7WxWoiEoMaVRDMrB1wEfB8rfA9wEVmtjJ8dg+Auy8BpgJLgVeBW9y95i5xNwOPEZloXgXMCPHHgXQzKwTuIJyx1NycmpPKVafn8MS7a1m7tSLa6YiINClrZl/EGy0/P98LCgpO+M8t2bWPL/7iLc7q04Xfj88/4T9fRORYmNkCd6/3j5euVD5CXVOSuOWCPsxcWsy7K7dGOx0RkSajgnAUvnV2Hj3SkvnJy0t0GqqIxAwVhKOQlBjPDy8dwKfFu3l63vpopyMi0iRUEI7SxQMzObtPOr94bQXbdldGOx0RkWOmgnCUzIwfXz6QPfur+cXrK6KdjojIMVNBOAZ9Mztyw9m5PDt/Aws37Ix2OiIix0QF4Rh9b1RfunRoy3+8tJiDB1vmKbwiIqCCcMw6JiXy/y7tz8KiMqYW6D5HItJyqSA0gbFDshmem8a9ry5nR8X+aKcjInJUVBCagJnx07GDKN9Xxb2vLo92OiIiR0UFoYn069aRiefk8ez8DSxY16xu1Coi0igqCE3oe6P60j01iR++sFhXMItIi6OC0ITat03gPy4fyPIt5fzx/bXRTkdE5IioIDSxiwdmckH/rtw/81M27twb7XRERBpNBaGJmRk/GTMQgB+9uJiWentxEWl9VBCOg5zO7bjjopOZvbyEVxZtiXY6IiKNooJwnHzzrFxOzU7lx39dQtlePYNZRJq/xj5Cs5OZTTOz5Wa2zMxGmlmamc00s5Vh2bnW9nebWaGZrTCzi2vFh5nZovDZg+HZyoTnLz8X4nPNLLfJe3qCJcTH8fMrT2Xb7krumaFrE0Sk+WvsEcKvgVfdvT9wGrCMyHOPZ7l7X2BWeI+ZDQDGAQOB0cAjZhYf9vMoMAnoG16jQ3wisMPd+wAPAPceY7+ahUHZqdx4bm+embeeOau2RTsdEZHDarAgmFkKcB7wOIC773f3ncAYYErYbAowNqyPAZ5190p3XwMUAsPNLAtIcfc5HplpfbJOm5p9TQNG1Rw9tHS3X3gyuentuOv5T9i7vzra6YiIfK7GHCH0BkqBJ8zsIzN7zMzaA5nuvhkgLLuG7bOB2nd5Kwqx7LBeN35IG3evAsqA9LqJmNkkMysws4LS0tJGdjG6ktvE8/MrB7Nu2x7un6nnJohI89WYgpAAnA486u5DgQrC8NDnqO+bvR8mfrg2hwbcJ7t7vrvnZ2RkHD7rZmTkSel8/cyePP7uGj7WcxNEpJlqTEEoAorcfW54P41IgSgOw0CEZUmt7XvUap8DbArxnHrih7QxswQgFYipGwLdfUl/MlOSuHPaQiqrNHQkIs1PgwXB3bcAG8ysXwiNApYC04EJITYBeCmsTwfGhTOH8ohMHs8Lw0rlZjYizA+Mr9OmZl9XA7M9xq7o6piUyM+uPJVPi3fz6zdWRjsdEZF/ktDI7W4F/mxmbYDVwA1EislUM5sIrAe+CuDuS8xsKpGiUQXc4u41X4lvBv4IJAMzwgsiE9ZPmVkhkSODccfYr2bpi/26ck1+Dr/9+yq+NLAbQ3p0inZKIiKfsZb6RTw/P98LCgqincYR27XvABc/8Dbt2ybw8q3nkJQY33AjEZEmYmYL3D2/vs90pfIJlpKUyD1XDaawZDcPvPFptNMREfmMCkIUfOHkDK4d3oPfv72agrUxNXcuIi2YCkKU/PDLA8junMwdUxeyu7Iq2umIiKggREuHtgncf80QNuzYw//8bWm00xERUUGIpjNy05h0Xm+embeBN5YWRzsdEWnlVBCi7I6LTqZ/t47c9fwnbN1dGe10RKQVU0GIsrYJ8fxq3BB27avizmmf6AlrIhI1KgjNQP9uKdx9SX9mLy/hTx+si3Y6ItJKqSA0E988K5cvnJzBf/9tGSuLy6Odjoi0QioIzYSZ8b9fHUyHtgl879mP2XdAN8ATkRNLBaEZ6doxif/96mCWbd6lx26KyAmngtDMXNA/k4nn5PHH99fy+pIt0U5HRFoRFYRm6M7R/Tg1O5V/m/YJm3bujXY6ItJKqCA0Q20T4nno2qFUVR/ktmc/oqr6YLRTEpFWQAWhmcrt0p6fXXkq89fu4JczdVdUETn+VBCasTFDsrl2eE8efWsVs5fr1hYicnypIDRz/3n5AE7JSuGOqQvZqPkEETmOGlUQzGytmS0ys4/NrCDE0sxsppmtDMvOtba/28wKzWyFmV1cKz4s7KfQzB4Mz1YmPH/5uRCfa2a5TdzPFispMZ5HrjudqmrnX57+kP1Vmk8QkePjSI4QvujuQ2o9eu0uYJa79wVmhfeY2QAiz0QeCIwGHjGzmudEPgpMAvqG1+gQnwjscPc+wAPAvUffpdiT16U99141mI/W7+RnryyLdjoiEqOOZchoDDAlrE8BxtaKP+vule6+BigEhptZFpDi7nM8cge3J+u0qdnXNGBUzdGDRHx5cNZn1ye8+NHGaKcjIjGosQXBgdfNbIGZTQqxTHffDBCWXUM8G9hQq21RiGWH9brxQ9q4exVQBqTXTcLMJplZgZkVlJaWNjL12HHXJf0ZnpvG3c8vYvmWXdFOR0RiTGMLwtnufjpwCXCLmZ13mG3r+2bvh4kfrs2hAffJ7p7v7vkZGRkN5RxzEuPjePjrQ+mQlMBNTy2gbO+BaKckIjGkUQXB3TeFZQnwAjAcKA7DQIRlSdi8COhRq3kOsCnEc+qJH9LGzBKAVEBPn69H15QkHrnudIp27OX25z7m4EE9P0FEmkaDBcHM2ptZx5p14EvAYmA6MCFsNgF4KaxPB8aFM4fyiEwezwvDSuVmNiLMD4yv06ZmX1cDs11PivlcZ+Sm8Z+XD2D28hIeeEMXrYlI00hoxDaZwAthjjcBeNrdXzWz+cBUM5sIrAe+CuDuS8xsKrAUqAJucfeaeznfDPwRSAZmhBfA48BTZlZI5MhgXBP0LaZdP6IXizfu4qHZhQzISuGSU7OinZKItHDWUr+I5+fne0FBQbTTiKrKqmrGTf6AFVvKef67Z9G/W0q0UxKRZs7MFtS6fOAQulK5BWubEM9vrx9Gh7YJ3DilgG27K6Odkoi0YCoILVxmShKTx+dTWl7JzX/SlcwicvRUEGLAkB6duO/qwcxbu50fvbiYljoMKCLR1ZhJZWkBxgzJprBkNw/NLqRvZgduPLd3tFMSkRZGBSGG3H7hyRSW7OZ/XllGz7R2fGlgt2inJCItiIaMYkhcnHH/NUMYnNOJ2579mEVFZdFOSURaEBWEGJPcJp7fjx9GWvs2TJwyX89kFpFGU0GIQV07JvGHb57B3v3V3PDEfHbt0z2PRKRhKggxql+3jvz2G8NYVbqbm55aoNNRRaRBKggx7Ow+Xbjv6sG8v2ob//6XT3Q6qogcls4yinFXnp7Dpp17+cXrn5KZksRdl/SPdkoi0kypILQCt3yxD5vL9vHbv68io2NbJp6TF+2URKQZUkFoBcyMn4wZxLbd+/npy0vp0qENY4ZkN9xQRFoVzSG0EvFxxq/GDeHMvDT+9f8W8vdPW98jSEXk8FQQWpGkxHh+PyGfPl07ctNTC1iwTg+lE5F/UEFoZVKSEnnyW8PJTGnLN5+Yz9JNu6Kdkog0EyoIrVBGx7b86cYz6dA2gfF/mMuarRXRTklEmoFGFwQzizezj8zs5fA+zcxmmtnKsOxca9u7zazQzFaY2cW14sPMbFH47MHwbGXC85efC/G5ZpbbhH2UeuR0bsdTE8/EHa77/Qds2L4n2imJSJQdyRHCbcCyWu/vAma5e19gVniPmQ0g8kzkgcBo4BEziw9tHgUmAX3Da3SITwR2uHsf4AHg3qPqjRyRPl078OTE4eyurOK6x+aypWxftFMSkShqVEEwsxzgy8BjtcJjgClhfQowtlb8WXevdPc1QCEw3MyygBR3n+ORS2afrNOmZl/TgFE1Rw9yfA3snsqTE89ke8V+vv7YB5SW6zGcIq1VY48QfgXcCdS+IU6mu28GCMuuIZ4NbKi1XVGIZYf1uvFD2rh7FVAGpNdNwswmmVmBmRWUluq0yaYypEcnnrjhDDbv3Md1j32gZzOLtFINFgQzuwwocfcFjdxnfd/s/TDxw7U5NOA+2d3z3T0/IyOjkelIY5yRm8bj38xn/fY9XPfYXBUFkVaoMUcIZwNfMbO1wLPABWb2J6A4DAMRliVh+yKgR632OcCmEM+pJ35IGzNLAFIBnSR/gp11Uhcen3AGa7ZWcN1jc9lesT/aKYnICdRgQXD3u909x91ziUwWz3b364HpwISw2QTgpbA+HRgXzhzKIzJ5PC8MK5Wb2YgwPzC+TpuafV0dfoZuzRkFZ/f5R1H4+u81fCTSmhzLdQj3ABeZ2UrgovAed18CTAWWAq8Ct7h7dWhzM5GJ6UJgFTAjxB8H0s2sELiDcMaSRMc5fSNFYe22CsZN/oCScp19JNIaWEv9Ip6fn+8FBQXRTiOmzVm1jYlT5tMtJYmnvz2CbqlJ0U5JRI6RmS1w9/z6PtOVyvK5Rp6UzpRvDaekvJKvTZ6ji9dEYpwKghzWGblpPDVxODsq9nPN7+awqnR3tFMSkeNEBUEaNLRnZ577zkgOVB/kmt/O0Q3xRGKUCoI0yilZKTz3nZG0SYjja5PnMH+tzgoWiTUqCNJoJ2V04P9uGklGh7Z84/G5zF5eHO2URKQJqSDIEcnp3I6pN42kT9cOfPvJBTz/YVHDjUSkRVBBkCPWpUNbnvn2CM7MS+OOqQv53d9X0VJPXxaRf1BBkKPSMSmRJ244gy8PzuLnM5bzk5eXcvCgioJIS5YQ7QSk5WqbEM9D44aS2TGJP7y3hpJdlfzymtNISoxvuLGINDsqCHJM4uKM/7h8AFmpSfzPK8so3rWPyePzSWvfJtqpicgR0pCRNIlvn9eb33z9dD7ZWMaVj7yn5zSLtEAqCNJkvjw4i2e+fSa79lVxxSPvMXf1tminJCJHQAVBmtSwXmm88N2zSGvfhusfn8vUgg0NNxKRZkEFQZpcr/T2vHDz2ZyZl86d0z7hZ68so1pnIIk0eyoIclyktouclvqNEb2Y/PZqbpwyn137DkQ7LRE5DBUEOW4S4+P46dhB/HTsIN5ZuZWxv3lPd0sVacZUEOS4+8aIXvzpxjPZuecAY3/znu6BJNJMNVgQzCzJzOaZ2UIzW2Jm/xXiaWY208xWhmXnWm3uNrNCM1thZhfXig8zs0XhswfDs5UJz19+LsTnmlnuceirRNGI3ulM/5ez6dG5HROnFPDrN1bqymaRZqYxRwiVwAXufhowBBhtZiOIPPd4lrv3BWaF95jZAGAcMBAYDTxiZjWXrj4KTAL6htfoEJ8I7HD3PsADwL3H3jVpbnI6t+P5757FFUOyeeCNT/n2kwWU7dW8gkhz0WBB8Iiagd/E8HJgDDAlxKcAY8P6GOBZd6909zVAITDczLKAFHef45E7oT1Zp03NvqYBo2qOHiS2JCXG88trTuMnYwby909Lufyhd1m8sSzaaYkIjZxDMLN4M/sYKAFmuvtcINPdNwOEZdeweTZQ++TzohDLDut144e0cfcqoAxIryePSWZWYGYFpaWljeqgND9mxviRuZ89he3KR9/n2XnrdcdUkShrVEFw92p3HwLkEPm2P+gwm9f3zd4PEz9cm7p5THb3fHfPz8jIaCBrae6G9erMy7eew5l5adz1/CJ+MHUhFZVV0U5LpNU6orOM3H0n8BaRsf/iMAxEWJaEzYqAHrWa5QCbQjynnvghbcwsAUgF9IzGViC9Q1v+eMNwbr/wZF78eCOXP/yuntksEiWNOcsow8w6hfVk4EJgOTAdmBA2mwC8FNanA+PCmUN5RCaP54VhpXIzGxHmB8bXaVOzr6uB2a7xg1YjPs647cK+/PnGEezeV8XYR97jqTlrNYQkcoI15gghC3jTzD4B5hOZQ3gZuAe4yMxWAheF97j7EmAqsBR4FbjF3avDvm4GHiMy0bwKmBHijwPpZlYI3EE4Y0lal5EnpfPKbecysnc6P3ppCZOeWsD2iv3RTkuk1bCW+i0sPz/fCwoKop2GHAcHDzp/eG8N9766nLT2bbj/miGc3adLtNMSiQlmtsDd8+v7TFcqS7MTF2fceG5vXvju2bRvm8B1j83lv19eyr4D1Q03FpGjpoIgzdag7FT+duu5fGNELx57dw1jf/MeyzZrwlnkeFFBkGYtuU08Px07iCduOINtFfv5ysPv8ps3C6mqPhjt1ERijgqCtAhf7NeV175/HhcP7Mb/vraCq347h8IS3TlVpCmpIEiLkda+DQ9//XQeunYo67ZVcOmD7/Dbv6/S0YJIE1FBkBbn8tO68/rt53FBv67cM2M5Vz36Piu2lEc7LZEWTwVBWqSuHZN49PrTefjrQ9mwYy+XPfQO97++gsoqnYkkcrRUEKTFMjMuG9ydmbefx2WDu/Pg7EIu/fU7zFuju56IHA0VBGnx0ju05YGvDWHKt4az78BBrvndHP592ifs0FXOIkdEBUFixhdOzmDmHefxnfN6M+3DIkbd/3emLSjSPZFEGkkFQWJKuzYJ3H3pKbx86znkprfjX/9vIdf8bo4uaBNpBBUEiUmnZKUw7aazuO+qwawqreCyh97lx9OX6JGdIoehgiAxKy7OuOaMHsz+wRe4dngPpsxZywW/eItn5q2n+qCGkUTqUkGQmNepXRv+e+yp/PVfzqF3Rnvufn4RX3n4Xeau3hbt1ESaFRUEaTUGZacy9TsjefDaoWyv2M/XJn/ATU8tYN22iminJtIsJEQ7AZETycz4ymndueiUTH7/zmoefWsVs5YXM35kLrde0IdO7dpEO0WRqNERgrRKyW3i+d6ovrz1b+dzxdBsnnhvDefd9ya/+/sqPXdBWq3GPFO5h5m9aWbLzGyJmd0W4mlmNtPMVoZl51pt7jazQjNbYWYX14oPM7NF4bMHw7OVCc9ffi7E55pZ7nHoq8g/yUxJ4r6rT+OV285lWK/O/HzGcr74i7eYOn+DbponrU5jjhCqgB+4+ynACOAWMxtA5LnHs9y9LzArvCd8Ng4YCIwGHjGz+LCvR4FJQN/wGh3iE4Ed7t4HeAC4twn6JtJo/bul8MQNw3n622fSNSWJO//yCRf/6m1eWbSZgzojSVqJBguCu2929w/DejmwDMgGxgBTwmZTgLFhfQzwrLtXuvsaoBAYbmZZQIq7z/HIpaNP1mlTs69pwKiaoweRE+msk7rw4nfP4rfXD8PM+O6fP+Syh95l1rJiXfEsMe+I5hDCUM5QYC6Q6e6bIVI0gK5hs2xgQ61mRSGWHdbrxg9p4+5VQBmQXs/Pn2RmBWZWUFpaeiSpizSamTF6UDde+/553H/NaeyurGLilALGPvI+b64oUWGQmNXogmBmHYC/AN9398PdB6C+b/Z+mPjh2hwacJ/s7vnunp+RkdFQyiLHJD7OuPL0HGb94Av8/MpT2VpeyQ1PzOeKR97nzeUqDBJ7GlUQzCyRSDH4s7s/H8LFYRiIsCwJ8SKgR63mOcCmEM+pJ35IGzNLAFIB3cNYmoXE+DiuHd6TN//1fH5+5amUlldywx/nc/nD7/Lq4i2aY5CY0ZizjAx4HFjm7vfX+mg6MCGsTwBeqhUfF84cyiMyeTwvDCuVm9mIsM/xddrU7OtqYLbr65c0M20S/lEY7rtqMOX7qrjpTwu4+Fdv8/yHRRzQWUnSwllDf3fN7BzgHWARUPMv/v8RmUeYCvQE1gNfdfftoc0PgW8ROUPp++4+I8TzgT8CycAM4FZ3dzNLAp4iMj+xHRjn7qsPl1d+fr4XFBQcaX9FmkxV9UFe/mQzj761ihXF5WR3SmbSeb35an4O7dromk9pnsxsgbvn1/tZS/0iroIgzcXBg87s5SU88lYhH67fSad2iYwf0YvxZ+XSpUPbaKcncggVBJETpGDtdn739mpmLi2mTUIcVwzJZuK5eZyc2THaqYkAhy8IOq4VaUL5uWnk56axqnQ3f3h3DX/5sIjnCjZwbt8u3HB2Luef3JW4OF1iI82TjhBEjqMdFfv589x1PPXBOop3VZKb3o7xI3O5algOqcmJ0U5PWiENGYlE2YHqg8xYvIUn3lvDR+t3kpwYz9ih2Ywf2YtTslKinZ60IioIIs3IoqIynvpgLS99vInKqoMM69WZ687syaWnZpGUGN/wDkSOgQqCSDO0c89+pi0o4um561m9tYLU5ESuGJrNuOE96N9NRw1yfKggiDRj7s4Hq7fz9Lz1vLZ4C/urDzKkRyeuye/BZadlkZKkuQZpOioIIi3E9or9vPDRRp6bv55Pi3eTlBjHJYOyuHpYDiN7p+sMJTlmKggiLYy780lRGVMLNjB94SbK91XRPTWJsUOzufL0bPp01XUNcnRUEERasH0Hqpm5tJjnPyzi7ZVbqT7oDMpOYeyQbC4/rTuZKUnRTlFaEBUEkRhRUr6Plxdu5sWPN/JJURlmMCIvna8M6c4lg7rRqV2baKcozZwKgkgMWlW6m+kfb+KvCzexemsFCXHGOX278OVTs/jSgG6kttNktPwzFQSRGObuLNm0i79+som/fbKZoh17SYgzzu7ThUtP7cZFA7qR1l5HDhKhgiDSSrg7C4vKmLFoM68s3syG7XuJMxiel8bogd340sBudO+UHO00JYpUEERaoZojh9eWbOHVxVtYWbIbgEHZKVx4SiYXnpLJwO4pRJ5XJa2FCoKIsKp0NzOXFvPG0mIWrN+BO2SlJnFB/66MOqUrI3t3IbmNbp0R61QQROQQW3dXMnt5CbOWFfPOyq3s2V9Nm4Q4RvZO54v9Mji/X1dyu7SPdppyHBxTQTCzPwCXASXuPijE0oDngFxgLXCNu+8In90NTASqge+5+2shPox/PD7zFeC28PjMtsCTwDBgG/A1d1/bUKdUEESaxr4D1cxbs503V5Tw5vIS1m7bA0Cv9HZ84eQMzu2bwYjeaXTULTRiwrEWhPOA3cCTtQrCfcB2d7/HzO4COrv7v5vZAOAZYDjQHXgDONndq81sHnAb8AGRgvCgu88ws+8Cg939JjMbB1zh7l9rqFMqCCLHx9qtFby9spS3VpQyZ9U29h6oJiHOGNqzE2f36cK5fbswOKcTifFx0U5VjsIxDxmZWS7wcq2CsAI43903m1kW8Ja79wtHB7j7z8N2rwE/JnIU8aa79w/xa0P779Rs4+5zzCwB2AJkeAOJqSCIHH+VVdV8uG4n76ws5Z2VW1m8qQx36NA2geF5aYzsnc7Ik9IZkJWi+yy1EMfjEZqZ7r4ZIBSFriGeTeQIoEZRiB0I63XjNW02hH1VmVkZkA5sracjk4BJAD179jzK1EWksdomxDPypMgf/TtHR27ZPWfVNt4t3Mqc1duYvbwEgJSkBIbnpTOidxpn5qVzSlZHEnQE0eI09TOV6/uK4IeJH67NPwfdJwOTIXKEcDQJisjR69SuDZecmsUlp2YBsKVsH3NWb2Xu6u18sHobbywrBiJHEMN6dWZ4Xhr5vTpzWo9OevhPC3C0BaHYzLJqDRmVhHgR0KPWdjnAphDPqSdeu01RGDJKBbYfZV4icgJ1S03iiqE5XDE08r/35rK9zFuz/bPX/762AoDEeGNQdir5vTozrFdnTu/Zma66KV+zc7QFYTowAbgnLF+qFX/azO4nMqncF5gXJpXLzWwEMBcYDzxUZ19zgKuB2Q3NH4hI85SVmsyYIdmMGRIZEd5RsZ8F63Ywf912Ply3gylz1vH7d9YAkN0pmaE9OzG0Z2eG9OjEwO4pOoqIssacZfQMcD7QBSgG/hN4EZgK9ATWA1919+1h+x8C3wKqgO+7+4wQz+cfp53OAG4Np50mAU8BQ4kcGYxz99UNJa5JZZGWp7KqmsUbd/HR+h18tH4nH67fweayfQAkxBn9szpyWk4nTsvpxOAeqfTJ6KC5iCamC9NEpNkq3rWPjzfs5OMNO1m4YSeLisoor6wCICkxjgFZKZyancrA7FQGdU+lb2YHnfJ6DFQQRKTFOHjQWbOtgkVFZXxSVMaijTtZsmkXe/ZXA9AmIY5+mR0Z2D2FAd1TOCUrhf7dOurCuUZSQRCRFq36oLN2WwWLN5axZNMulm7axZJNZezYc+CzbXqmtaNft46c0q0j/bql0K9bR3LT22nIqY7jcR2CiMgJEx9nnJTRgZMyOnw2Ye3ubNm1j2Wbd7FsczlLN+9ixZZyZi0r5mD4ntsmIY4+GR04ObMDfTM7cnJmR/p27UCPtHbE60K6f6KCICItkpmRlZpMVmoyF/TP/Cy+70A1hSW7WbGlnBXF5azYUs68Ndt58eNNn23TJiGO3l3a06drh89eJ2V0IK9L+1Z9ppMKgojElKTEeAZlpzIoO/WQ+K59B1hZvJtVJbspLN3NyuJyFhbt5G+LNlMzcm4G3VOT6Z3Rnt5d2pPXpT25YZndKTnmh59UEESkVUhJSmRYuDCutn0HqlldWsHqrbtZXVrBqtLdrNlawV8+3MjucLYTRC6u69G5Hb3S29ErvX1YtqNnWntyOifHxJGFCoKItGpJifEMCGcs1ebulO6uZO3WPazdWsGabRWs21bB2q17mLdmOxXhrCeIHFlkdkyiZ1o7ctKS6dG5HT3S2tGjczLZnZPplpLUIo4uVBBEROphZnTtmETXjkkMz0s75DN3Z1vFftZt28P67RWs37aX9dsj6+8XbqO4fCO1T+CMjzO6pSSR3TmZnE6RItG9U+SV3SmJrNRk2reN/p/j6GcgItLCmBldOrSlS4e2/zQEBZErsjft3MeG7XvYuHMvG3fspWhHZH3umu1s/njvZ2dC1UhJSqB7p2S6pSaFyfIkuqUkkZmaRFZqEpkpSaQkJRzXZ2CrIIiINLG2CfHkhcno+lRVH6S4vJJNoVhsLtvH5rK9bNq5jy279rKoqIxtFfv/qV1SYhyZKUn84Ev9+Mpp3Zs8bxUEEZETLCE+juxOyWR3SuaM3Pq3qayqpmRXJVt27WNz2T5Kdu2jeNc+tuyqJL19m+OT13HZq4iIHJO2CfGRiem0difsZzb/aW8RETkhVBBERARQQRARkUAFQUREABUEEREJVBBERARQQRARkUAFQUREgBb8CE0zKwXWHWXzLsDWJkynJVCfWwf1uXU4lj73cveM+j5osQXhWJhZwec9UzRWqc+tg/rcOhyvPmvISEREABUEEREJWmtBmBztBKJAfW4d1OfW4bj0uVXOIYiIyD9rrUcIIiJShwqCiIgArbAgmNloM1thZoVmdle082kKZtbDzN40s2VmtsTMbgvxNDObaWYrw7JzrTZ3h9/BCjO7OHrZHxszizezj8zs5fA+pvtsZp3MbJqZLQ//vUe2gj7fHv5dLzazZ8wsKdb6bGZ/MLMSM1tcK3bEfTSzYWa2KHz2oB3pA5jdvdW8gHhgFdAbaAMsBAZEO68m6FcWcHpY7wh8CgwA7gPuCvG7gHvD+oDQ97ZAXvidxEe7H0fZ9zuAp4GXw/uY7jMwBbgxrLcBOsVyn4FsYA2QHN5PBb4Za30GzgNOBxbXih1xH4F5wEjAgBnAJUeSR2s7QhgOFLr7anffDzwLjIlyTsfM3Te7+4dhvRxYRuR/pDFE/oAQlmPD+hjgWXevdPc1QCGR302LYmY5wJeBx2qFY7bPZpZC5A/H4wDuvt/ddxLDfQ4SgGQzSwDaAZuIsT67+9vA9jrhI+qjmWUBKe4+xyPV4clabRqltRWEbGBDrfdFIRYzzCwXGArMBTLdfTNEigbQNWwWK7+HXwF3AgdrxWK5z72BUuCJMEz2mJm1J4b77O4bgV8A64HNQJm7v04M97mWI+1jdlivG2+01lYQ6htPi5nzbs2sA/AX4Pvuvutwm9YTa1G/BzO7DChx9wWNbVJPrEX1mcg35dOBR919KFBBZCjh87T4Podx8zFEhka6A+3N7PrDNakn1qL63Aif18dj7ntrKwhFQI9a73OIHH62eGaWSKQY/Nndnw/h4nAYSViWhHgs/B7OBr5iZmuJDP1dYGZ/Irb7XAQUufvc8H4akQIRy32+EFjj7qXufgB4HjiL2O5zjSPtY1FYrxtvtNZWEOYDfc0sz8zaAOOA6VHO6ZiFMwkeB5a5+/21PpoOTAjrE4CXasXHmVlbM8sD+hKZjGox3P1ud89x91wi/x1nu/v1xHaftwAbzKxfCI0ClhLDfSYyVDTCzNqFf+ejiMyRxXKfaxxRH8OwUrmZjQi/q/G12jROtGfXozCbfymRs3BWAT+Mdj5N1KdziBwafgJ8HF6XAunALGBlWKbVavPD8DtYwRGeidDcXsD5/OMso5juMzAEKAj/rV8EOreCPv8XsBxYDDxF5OyamOoz8AyROZIDRL7pTzyaPgL54fe0CniYcDeKxr506woREQFa35CRiIh8DhUEEREBVBBERCRQQRAREUAFQUREAhUEEREBVBBERCT4/0NLB0wp18xOAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "xx = numpy.linspace(0,1000,100)\n",
    "yy = 80000*numpy.exp(-0.003*xx)\n",
    "plt.plot(xx,yy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e676875",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab11e9e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e8562bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "164543f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "293d6af9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b721e62",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "737039e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "siteuse_rea 331113\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import statistics\n",
    "import msprime\n",
    "from IPython.display import SVG, display\n",
    "import tsinfer\n",
    "import scipy\n",
    "import math\n",
    "import numpy\n",
    "import tskit\n",
    "import io\n",
    "import builtins\n",
    "import sys\n",
    "from tqdm.notebook import tqdm\n",
    "from tskit import MISSING_DATA\n",
    "import pickle\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import cvxopt as opt\n",
    "from cvxopt import matrix, spmatrix, sparse\n",
    "from cvxopt.solvers import qp, options\n",
    "from cvxopt import blas\n",
    "from cvxopt import spdiag\n",
    "import statsmodels\n",
    "from statsmodels.nonparametric.kernel_regression import KernelReg\n",
    "from scipy.signal import savgol_filter\n",
    "import time\n",
    "import bisect\n",
    "\n",
    "\n",
    "def find_s(elem, sorted_list):\n",
    "    'Locate the leftmost value exactly equal to x'\n",
    "    i = bisect.bisect_left(sorted_list, elem)\n",
    "    if i != len(sorted_list) and sorted_list[i] == elem:\n",
    "        return i\n",
    "\n",
    "\n",
    "def find_s_se(elem, sorted_list):\n",
    "    'Locate the leftmost value smaller or equal to x'\n",
    "    i = bisect.bisect_right(sorted_list, elem)\n",
    "    if i != 0:\n",
    "        return i-1\n",
    "    else:\n",
    "        print(\"error\")\n",
    "\n",
    "\n",
    "def find_s_s(elem, sorted_list):\n",
    "    'Locate the leftmost value smaller or equal to x'\n",
    "    i = bisect.bisect_left(sorted_list, elem)\n",
    "    if i != 0:\n",
    "        return i\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    " \n",
    "    \n",
    "\n",
    "    \n",
    "file = open(\"recombination_map.dat\",\"rb\")\n",
    "rmap = pickle.load(file)\n",
    "file.close\n",
    "\n",
    "sam=tsinfer.load(\"tgp_chr20.samples\")\n",
    "\n",
    "LWKlist = []\n",
    "for i in range(len(sam.individuals_population)):\n",
    "    if sam.individuals_population[i]==11:\n",
    "        LWKlist.append(i)\n",
    "newdata = sam.subset(individuals=LWKlist)\n",
    "\n",
    "        \n",
    "    \n",
    "    \n",
    "seq_len = 63025522\n",
    "sam_size = 99\n",
    "r = rmap\n",
    "mu = 1.375*10**(-8)\n",
    "# gene_conver = 10**(-8)\n",
    "# track_len = 100\n",
    "# epsilon = 0.0001\n",
    "totaliter=1\n",
    "\n",
    "simtime = 2\n",
    "# priormu = [1.0*10**(-8), 2.0*10**(-8)]\n",
    "# prioreps = [0, 0.0002]\n",
    "# accept_threshold = [1000,400]\n",
    "freqconsider = [1,300]\n",
    "priorinipop = [120000, 5000]\n",
    "priorinirate = [-0.001 , 0.004]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "inipopsimlist = []\n",
    "iniratesimlist = []\n",
    "summary_sta1_simlist = []\n",
    "summary_sta2_simlist = []\n",
    "\n",
    "current_pop = []\n",
    "for iter in range(totaliter):\n",
    "\n",
    "    sitefreq = numpy.sum(newdata.sites_genotypes,1)\n",
    "    siteuse =  numpy.where( (sitefreq>freqconsider[0] ) * ( sitefreq<freqconsider[1] ) )[0]  \n",
    "    print(\"siteuse_rea\",len(siteuse))\n",
    "    datause = newdata.subset(sites = siteuse)\n",
    "    inferred_ts = tsinfer.infer(datause,path_compression=False)\n",
    "\n",
    "#     inferred_ts = tsinfer.infer(newdata,path_compression=False)\n",
    "\n",
    "\n",
    "\n",
    "    mts = inferred_ts\n",
    "\n",
    "    ###############################################################################\n",
    "    msites = mts.tables.mutations.site\n",
    "    sitepos = mts.tables.sites.position\n",
    "    mpos = []\n",
    "    for i in range(len(msites)):\n",
    "        cursite = msites[i]\n",
    "        mpos.append(sitepos[cursite])\n",
    "    mpos = numpy.array(mpos)\n",
    "    mnodes = mts.tables.mutations.node\n",
    "    ec = mts.tables.edges.child\n",
    "    ep = mts.tables.edges.parent\n",
    "    el = mts.tables.edges.left\n",
    "    er = mts.tables.edges.right\n",
    "\n",
    "\n",
    "    leftmost = min(el)\n",
    "    rightmost = max(er)\n",
    "    for i in range(len(el)):\n",
    "        if el[i] == leftmost:\n",
    "            el[i] = 0\n",
    "        if er[i] == rightmost:\n",
    "            er[i] = seq_len\n",
    "\n",
    "\n",
    "    if len(mpos) == 0:\n",
    "        mpos_sn = []\n",
    "        mnodes_sn = []\n",
    "    else:\n",
    "        mnp_sn = numpy.transpose(sorted(numpy.transpose(numpy.array([mnodes,mpos])).tolist()))\n",
    "        mnodes_sn = mnp_sn[0]\n",
    "        mpos_sn = mnp_sn[1]\n",
    "\n",
    "    eclrp_sc = numpy.transpose(sorted(numpy.transpose(numpy.array([ec,el,er,ep])).tolist()))\n",
    "    ec_sc = eclrp_sc[0]\n",
    "    el_sc = eclrp_sc[1]\n",
    "    er_sc = eclrp_sc[2]\n",
    "    ep_sc = eclrp_sc[3]\n",
    "    enm_sc = [[] for i in range(len(ec_sc))]\n",
    "\n",
    "    curedge = 0\n",
    "    for i in range(len(mnodes_sn)):\n",
    "        curnodemu = mnodes_sn[i] \n",
    "        cursitemu = mpos_sn[i]\n",
    "        for j in range(curedge,len(ec_sc)):\n",
    "            if curnodemu == ec_sc[j]:\n",
    "\n",
    "                if er_sc[j] > cursitemu and el_sc[j] <= cursitemu:\n",
    "\n",
    "                    enm_sc[j].append(cursitemu)  \n",
    "\n",
    "                    curedge = j\n",
    "                    break\n",
    "                if el_sc[j] > cursitemu:\n",
    "                    curedge = j\n",
    "                    break\n",
    "\n",
    "            if ec_sc[j] > mnodes_sn[i] or j == len(ec_sc)-1:\n",
    "                #print(\"missing mutations\")\n",
    "                curedge = j\n",
    "                break\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ###########################################################################\n",
    "    ###########################################################################\n",
    "    ##########################################################################\n",
    "\n",
    "    clocate = []\n",
    "    curc = -1\n",
    "    for i in range(len(ec_sc)):\n",
    "        while ec_sc[i] != curc:\n",
    "            curc=curc+1\n",
    "            clocate.append(i)\n",
    "    clocate.append(len(ec_sc))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    abdlist = []\n",
    "\n",
    "    for i in range(0,2*sam_size-1):\n",
    "        for j in range(i+1,i+2):\n",
    "            canabd = []\n",
    "            curleft = 0\n",
    "            curright =seq_len\n",
    "            #start = 0\n",
    "            while curleft != seq_len:\n",
    "                curp1 = i\n",
    "                curp2 = j \n",
    "                curmed = []\n",
    "                while curp1 != curp2:\n",
    "                    if curp1 < curp2:\n",
    "                        begin1 = clocate[curp1]\n",
    "                        end = clocate[curp1+1]\n",
    "                        s_range = el_sc[begin1:end]\n",
    "                        begin2 = find_s_se(curleft,s_range)\n",
    "                        for k in range(begin1+begin2,end):\n",
    "                            if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                curp1 = int(ep_sc[k])\n",
    "                                curoldr = curright\n",
    "                                curright = min(er_sc[k],curright)\n",
    "\n",
    "                                curmed.append(k)\n",
    "\n",
    "                                break\n",
    "\n",
    "                    else:\n",
    "                        begin1 = clocate[curp2]\n",
    "                        end = clocate[curp2+1]\n",
    "                        s_range = el_sc[begin1:end]\n",
    "                        begin2 = find_s_se(curleft,s_range)\n",
    "                        for k in range(begin1+begin2,end):\n",
    "                            if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                curp2 = int(ep_sc[k])\n",
    "                                curoldr = curright\n",
    "                                curright = min(er_sc[k],curright) \n",
    "\n",
    "                                curmed.append(k)\n",
    "\n",
    "                                break\n",
    "\n",
    "                canmution = []\n",
    "                for k in range(len(curmed)):\n",
    "                    edgemu = enm_sc[curmed[k]]\n",
    "                    startt = find_s_s(curleft,edgemu)\n",
    "                    endd = find_s_s(curright,edgemu)                    \n",
    "                    canmution = canmution + edgemu[startt:endd] \n",
    "\n",
    "                canabd.append([i,j,curleft,curright,curp1,canmution])\n",
    "                curleft = curright\n",
    "                curright = seq_len\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            if len(canabd)==1:\n",
    "                abdlist = abdlist + canabd\n",
    "            else:\n",
    "                curabd = [canabd[0][0],canabd[0][1],canabd[0][2],canabd[0][3],canabd[0][4],canabd[0][5]]\n",
    "                for k in range(1,len(canabd)):\n",
    "                    if canabd[k][4] != canabd[k-1][4]:\n",
    "                        abdlist.append(curabd)\n",
    "                        curabd = [canabd[k][0],canabd[k][1],canabd[k][2],canabd[k][3],canabd[k][4],canabd[k][5]]\n",
    "                    else:\n",
    "                        curabd[3] = canabd[k][3]\n",
    "                        curabd[5] = curabd[5] + canabd[k][5]\n",
    "                    if k == len(canabd)-1:\n",
    "                        abdlist.append(curabd)\n",
    "\n",
    "\n",
    "\n",
    "#             print(\"abdlist sample:\",i)\n",
    "    print(\"numibd\",len(abdlist))\n",
    "    \n",
    "    abduse = abdlist\n",
    "    \n",
    "##############################    \n",
    "    \n",
    "#     tolnummu = 0\n",
    "#     for i in range(len(abduse)):\n",
    "#         tolnummu = tolnummu+len(abduse[i][5])\n",
    "\n",
    "#     muloclist = np.zeros(tolnummu)\n",
    "#     indd = 0\n",
    "#     for i in range(len(abduse)):\n",
    "#         abdc = abduse[i]\n",
    "#         lee = len(abdc[5])\n",
    "#         muloclist[indd:(indd+lee)] = abdc[5]\n",
    "#         indd = indd+lee\n",
    "\n",
    "#     smulist = numpy.sort(muloclist)\n",
    "#     count2 = 0\n",
    "#     curloc = -1\n",
    "#     curcount = 1\n",
    "#     for i in range(len(smulist)):\n",
    "#         if smulist[i] == curloc:\n",
    "#             curcount = curcount+1\n",
    "#         else:\n",
    "#             if curcount ==2:\n",
    "#                 count2 = count2+1\n",
    "#             curcount = 1\n",
    "#             curloc = smulist[i]\n",
    "\n",
    "# #     summary_sta1_inf = (len(smulist) - 2*count2)/len(abduse)        \n",
    "# #     summary_sta2_inf = (2*count2) / len(abduse)  \n",
    "\n",
    "\n",
    "#     summary_sta1_inf = (len(smulist) - 2*count2)      \n",
    "#     summary_sta2_inf = (2*count2)         \n",
    "\n",
    "####################################################################\n",
    "\n",
    "    lenlist = numpy.zeros(len(abduse))\n",
    "    for i in range(len(abduse)):\n",
    "        lenlist[i] = abduse[i][3] - abduse[i][2]\n",
    "\n",
    "    summary_sta1_inf = statistics.mean(lenlist)\n",
    "    summary_sta2_inf = statistics.stdev(lenlist)\n",
    "\n",
    "#################################################################################################\n",
    "    stime = time.time()\n",
    "\n",
    "    for run in range(simtime):\n",
    "        \n",
    "        popsim = numpy.random.uniform(priorinipop[0],priorinipop[1])\n",
    "        ratesim = numpy.random.uniform(priorinirate[0],priorinirate[1])\n",
    "\n",
    "        if run == 0:\n",
    "            popsim = 13000\n",
    "            ratesim = 0\n",
    "        if run == 1:\n",
    "            popsim = 20000\n",
    "            ratesim = 0\n",
    "#         if run == 2:\n",
    "#             popsim = 8000\n",
    "#             ratesim = 0.001\n",
    "#         if run == 3:\n",
    "#             popsim = 8000\n",
    "#             ratesim = 0.004            \n",
    "#         if run == 4:\n",
    "#             popsim = 20000\n",
    "#             ratesim = 0.000          \n",
    "#         if run == 5:\n",
    "#             popsim = 20000\n",
    "#             ratesim = 0.000    \n",
    "        \n",
    "        \n",
    "###      model  pop = const_a + inip * exp(inir * x)   when x<1000, when x =1000, pop = known pop, 2 parameters          \n",
    "        inipopsimlist.append(popsim)\n",
    "        iniratesimlist.append(ratesim)\n",
    "        \n",
    "        \n",
    "#         demo_model = msprime.Demography.isolated_model([popsim], growth_rate=[ratesim])\n",
    "#         demo_model.add_population_parameters_change(1000,  initial_size=None, growth_rate=0, population=None)\n",
    "#         if ratesim < 0.0001:\n",
    "#             demo_model.add_population_parameters_change(2000,  initial_size=None, growth_rate=0.0001, population=None)\n",
    "        \n",
    "        \n",
    "\n",
    "        genpoint = [0*5/5.2,400*5/5.2,800*5/5.2,1200*5/5.2,1600*5/5.2,2000*5/5.2,2400*5/5.2,2800*5/5.2,\\\n",
    "                    3200*5/5.2,3600*5/5.2,4000*5/5.2,8000*5/5.2,12000*5/5.2,16000*5/5.2,20000*5/5.2]\n",
    "        inipopsize =[15200*5/5.2, 15200*5/5.2, 15600*5/5.2, 14800*5/5.2, 12400*5/5.2, 10400*5/5.2,9800*5/5.2,\\\n",
    "                     10400*5/5.2,12000*5/5.2,14400*5/5.2,16800*5/5.2,29400*5/5.2,28000*5/5.2,22000*5/5.2,18000*5/5.2]\n",
    "        genuse = []\n",
    "        inipuse = []\n",
    "        gruse = []\n",
    "        for i in range(len(genpoint)-1):\n",
    "            addplist = []\n",
    "            for j in range(10):\n",
    "                addplist.append(genpoint[i] + (genpoint[i+1]-genpoint[i])/10*j)\n",
    "            genuse = genuse + addplist               \n",
    "        for i in range(len(inipopsize)-1):\n",
    "            addplist = []\n",
    "            for j in range(10):\n",
    "                addplist.append(inipopsize[i] + (inipopsize[i+1]-inipopsize[i])/10*j)\n",
    "            inipuse = inipuse + addplist\n",
    "        genuse.append(genpoint[len(genpoint)-1])\n",
    "        inipuse.append(inipopsize[len(inipopsize)-1])\n",
    "\n",
    "        \n",
    "#         ind1000 = 0\n",
    "#         for i in range(len(genuse)):\n",
    "#             if genuse[i]<1000:\n",
    "#                 ind1000 = i\n",
    "#         consta = inipuse[ind1000] - popsim*numpy.exp(-ratesim*1000)        \n",
    "#         for i in range(len(genuse)):\n",
    "#             if genuse[i]<1000:\n",
    "#                 inipuse[i] = consta + popsim*numpy.exp(-ratesim*genuse[i]) \n",
    "#         for i in range(len(genuse)-1):\n",
    "#             gruse.append( numpy.log( inipuse[i]/inipuse[i+1] )/(genuse[i+1] - genuse[i])  )\n",
    "#         gruse.append(0)\n",
    "#         \n",
    "#         current_pop.append( consta + popsim)\n",
    "#         if consta+popsim <1000:\n",
    "#             demo_model = msprime.Demography.isolated_model([1000], growth_rate=[0.001])\n",
    "#         else:\n",
    "#             demo_model = msprime.Demography.isolated_model([inipuse[0]], growth_rate=[gruse[0]])\n",
    "#             for i in range(1,len(genuse)):\n",
    "#                 demo_model.add_population_parameters_change(genuse[i],  initial_size=inipuse[i], growth_rate=gruse[i], population=None)\n",
    "\n",
    "        for i in range(len(genuse)-1):\n",
    "            gruse.append( numpy.log( inipuse[i]/inipuse[i+1] )/(genuse[i+1] - genuse[i])  )\n",
    "        gruse.append(0)\n",
    "        \n",
    "        demo_model = msprime.Demography.isolated_model([inipuse[0]], growth_rate=[gruse[0]])\n",
    "        for i in range(1,len(genuse)):\n",
    "            demo_model.add_population_parameters_change(genuse[i],  initial_size=inipuse[i], growth_rate=gruse[i], population=None)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#####################################################\n",
    "        ts = msprime.sim_ancestry(\n",
    "            samples=sam_size,\n",
    "            recombination_rate= r, \n",
    "            sequence_length= seq_len,\n",
    "            record_provenance=False,\n",
    "#             population_size = pop_size,\n",
    "#             gene_conversion_rate = gene_conver ,\n",
    "#             gene_conversion_tract_length = track_len,\n",
    "            #random_seed =seed,\n",
    "            #discrete_genome=False\n",
    "            demography = demo_model\n",
    "            )\n",
    "        # Visualise the simulated ancestral history.\n",
    "        #SVG(ts.draw_svg())\n",
    "\n",
    "        #ts.num_trees\n",
    "        \n",
    "        print(\"done gene\")\n",
    "\n",
    "        mts = msprime.sim_mutations(ts, rate = mu,\n",
    "                                    #discrete_genome=False,\n",
    "                                    #random_seed=seed\n",
    "                                  )\n",
    "\n",
    "        print(\"done mu\")\n",
    "\n",
    "        sample_data = tsinfer.SampleData.from_tree_sequence(mts, use_sites_time=None, use_individuals_time=None)\n",
    "\n",
    "    \n",
    "        sitefreq = numpy.sum(sample_data.sites_genotypes,1)\n",
    "        siteuse =  numpy.where( (sitefreq>freqconsider[0] ) * ( sitefreq<freqconsider[1] ) )[0]  \n",
    "        print(\"siteuse_sim\",len(siteuse))\n",
    "        datause = sample_data.subset(sites = siteuse)\n",
    "        inferred_ts = tsinfer.infer(datause,path_compression=False)\n",
    "\n",
    "#         inferred_ts = tsinfer.infer(sample_data,path_compression=False)\n",
    "\n",
    "        mts = inferred_ts\n",
    "\n",
    "        ###############################################################################\n",
    "        msites = mts.tables.mutations.site\n",
    "        sitepos = mts.tables.sites.position\n",
    "        mpos = []\n",
    "        for i in range(len(msites)):\n",
    "            cursite = msites[i]\n",
    "            mpos.append(sitepos[cursite])\n",
    "        mpos = numpy.array(mpos)\n",
    "        mnodes = mts.tables.mutations.node\n",
    "        ec = mts.tables.edges.child\n",
    "        ep = mts.tables.edges.parent\n",
    "        el = mts.tables.edges.left\n",
    "        er = mts.tables.edges.right\n",
    "\n",
    "\n",
    "        leftmost = min(el)\n",
    "        rightmost = max(er)\n",
    "        for i in range(len(el)):\n",
    "            if el[i] == leftmost:\n",
    "                el[i] = 0\n",
    "            if er[i] == rightmost:\n",
    "                er[i] = seq_len\n",
    "\n",
    "\n",
    "        if len(mpos) == 0:\n",
    "            mpos_sn = []\n",
    "            mnodes_sn = []\n",
    "        else:\n",
    "            mnp_sn = numpy.transpose(sorted(numpy.transpose(numpy.array([mnodes,mpos])).tolist()))\n",
    "            mnodes_sn = mnp_sn[0]\n",
    "            mpos_sn = mnp_sn[1]\n",
    "\n",
    "        eclrp_sc = numpy.transpose(sorted(numpy.transpose(numpy.array([ec,el,er,ep])).tolist()))\n",
    "        ec_sc = eclrp_sc[0]\n",
    "        el_sc = eclrp_sc[1]\n",
    "        er_sc = eclrp_sc[2]\n",
    "        ep_sc = eclrp_sc[3]\n",
    "        enm_sc = [[] for i in range(len(ec_sc))]\n",
    "\n",
    "        curedge = 0\n",
    "        for i in range(len(mnodes_sn)):\n",
    "            curnodemu = mnodes_sn[i] \n",
    "            cursitemu = mpos_sn[i]\n",
    "            for j in range(curedge,len(ec_sc)):\n",
    "                if curnodemu == ec_sc[j]:\n",
    "\n",
    "                    if er_sc[j] > cursitemu and el_sc[j] <= cursitemu:\n",
    "\n",
    "                        enm_sc[j].append(cursitemu)  \n",
    "\n",
    "                        curedge = j\n",
    "                        break\n",
    "                    if el_sc[j] > cursitemu:\n",
    "                        curedge = j\n",
    "                        break\n",
    "\n",
    "                if ec_sc[j] > mnodes_sn[i] or j == len(ec_sc)-1:\n",
    "                    #print(\"missing mutations\")\n",
    "                    curedge = j\n",
    "                    break\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ###########################################################################\n",
    "        ###########################################################################\n",
    "        ##########################################################################\n",
    "\n",
    "        clocate = []\n",
    "        curc = -1\n",
    "        for i in range(len(ec_sc)):\n",
    "            while ec_sc[i] != curc:\n",
    "                curc=curc+1\n",
    "                clocate.append(i)\n",
    "        clocate.append(len(ec_sc))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        abdlist = []\n",
    "\n",
    "        for i in range(0,2*sam_size-1):\n",
    "            for j in range(i+1,i+2):\n",
    "                canabd = []\n",
    "                curleft = 0\n",
    "                curright =seq_len\n",
    "                #start = 0\n",
    "                while curleft != seq_len:\n",
    "                    curp1 = i\n",
    "                    curp2 = j \n",
    "                    curmed = []\n",
    "                    while curp1 != curp2:\n",
    "                        if curp1 < curp2:\n",
    "                            begin1 = clocate[curp1]\n",
    "                            end = clocate[curp1+1]\n",
    "                            s_range = el_sc[begin1:end]\n",
    "                            begin2 = find_s_se(curleft,s_range)\n",
    "                            for k in range(begin1+begin2,end):\n",
    "                                if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                    curp1 = int(ep_sc[k])\n",
    "                                    curoldr = curright\n",
    "                                    curright = min(er_sc[k],curright)\n",
    "\n",
    "                                    curmed.append(k)\n",
    "\n",
    "                                    break\n",
    "\n",
    "                        else:\n",
    "                            begin1 = clocate[curp2]\n",
    "                            end = clocate[curp2+1]\n",
    "                            s_range = el_sc[begin1:end]\n",
    "                            begin2 = find_s_se(curleft,s_range)\n",
    "                            for k in range(begin1+begin2,end):\n",
    "                                if el_sc[k]<=curleft and er_sc[k] > curleft:\n",
    "                                    curp2 = int(ep_sc[k])\n",
    "                                    curoldr = curright\n",
    "                                    curright = min(er_sc[k],curright) \n",
    "\n",
    "                                    curmed.append(k)\n",
    "\n",
    "                                    break\n",
    "\n",
    "                    canmution = []\n",
    "                    for k in range(len(curmed)):\n",
    "                        edgemu = enm_sc[curmed[k]]\n",
    "                        startt = find_s_s(curleft,edgemu)\n",
    "                        endd = find_s_s(curright,edgemu)                    \n",
    "                        canmution = canmution + edgemu[startt:endd] \n",
    "\n",
    "                    canabd.append([i,j,curleft,curright,curp1,canmution])\n",
    "                    curleft = curright\n",
    "                    curright = seq_len\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "                if len(canabd)==1:\n",
    "                    abdlist = abdlist + canabd\n",
    "                else:\n",
    "                    curabd = [canabd[0][0],canabd[0][1],canabd[0][2],canabd[0][3],canabd[0][4],canabd[0][5]]\n",
    "                    for k in range(1,len(canabd)):\n",
    "                        if canabd[k][4] != canabd[k-1][4]:\n",
    "                            abdlist.append(curabd)\n",
    "                            curabd = [canabd[k][0],canabd[k][1],canabd[k][2],canabd[k][3],canabd[k][4],canabd[k][5]]\n",
    "                        else:\n",
    "                            curabd[3] = canabd[k][3]\n",
    "                            curabd[5] = curabd[5] + canabd[k][5]\n",
    "                        if k == len(canabd)-1:\n",
    "                            abdlist.append(curabd)\n",
    "\n",
    "\n",
    "    #             print(\"abdlist sample:\",i)\n",
    "        print(\"numibd\",len(abdlist))\n",
    "\n",
    "        abduse = abdlist\n",
    "\n",
    "##########################################################################\n",
    "\n",
    "#         tolnummu = 0\n",
    "#         for i in range(len(abduse)):\n",
    "#             tolnummu = tolnummu+len(abduse[i][5])\n",
    "\n",
    "\n",
    "#         muloclist = np.zeros(tolnummu)\n",
    "#         indd = 0\n",
    "#         for i in range(len(abduse)):\n",
    "#             abdc = abduse[i]\n",
    "#             lee = len(abdc[5])\n",
    "#             muloclist[indd:(indd+lee)] = abdc[5]\n",
    "#             indd = indd+lee\n",
    "            \n",
    "#         smulist = numpy.sort(muloclist)\n",
    "#         count2 = 0\n",
    "#         curloc = -1\n",
    "#         curcount = 1\n",
    "#         for i in range(len(smulist)):\n",
    "#             if smulist[i] == curloc:\n",
    "#                 curcount = curcount+1\n",
    "#             else:\n",
    "#                 if curcount ==2:\n",
    "#                     count2 = count2+1\n",
    "#                 curcount = 1\n",
    "#                 curloc = smulist[i]\n",
    "        \n",
    "# #         summary_sta1_sim = (len(smulist) - 2*count2)/len(abduse)        \n",
    "# #         summary_sta2_sim = (2*count2) / len(abduse)\n",
    "\n",
    "#         summary_sta1_sim = (len(smulist) - 2*count2)        \n",
    "#         summary_sta2_sim = (2*count2)  \n",
    "\n",
    "#         summary_sta1_simlist.append(summary_sta1_sim)\n",
    "#         summary_sta2_simlist.append(summary_sta2_sim)\n",
    "        \n",
    "        \n",
    "#########################################################################        \n",
    "        \n",
    "        lenlist = numpy.zeros(len(abduse))\n",
    "        for i in range(len(abduse)):\n",
    "            lenlist[i] = abduse[i][3] - abduse[i][2]\n",
    "\n",
    "        summary_sta1_simlist.append( statistics.mean(lenlist) )\n",
    "        summary_sta2_simlist.append( statistics.stdev(lenlist) )\n",
    "        \n",
    "        print(\"-----------------------------------\")\n",
    "        print(inipopsimlist )\n",
    "        print(iniratesimlist )\n",
    "        print(summary_sta1_simlist,summary_sta1_inf)\n",
    "        print(summary_sta2_simlist,summary_sta2_inf)\n",
    "        print(time.time()-stime, run)\n",
    "        print(\"-----------------------------------\")\n",
    "\n",
    "\n",
    "\n",
    "# byrun = 1\n",
    "# name1 = \"b_ne_realpopsimall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "# name2 = \"b_ne_realratesimall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "# name3 = \"b_ne_realsta1simall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "# name4 = \"b_ne_realsta2simall_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "# name5 = \"b_ne_realsta1inf_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "# name6 = \"b_ne_realsta2inf_\"+\"run=\"+str(byrun) +\"LWK_efibd.dat\"\n",
    "\n",
    "# file = open(name1,\"wb\") \n",
    "# pickle.dump(inipopsimlist,file)\n",
    "# file.close()\n",
    "\n",
    "# file = open(name2,\"wb\") \n",
    "# pickle.dump(iniratesimlist,file)\n",
    "# file.close()\n",
    "\n",
    "# file = open(name3,\"wb\") \n",
    "# pickle.dump(summary_sta1_simlist,file)\n",
    "# file.close()\n",
    "\n",
    "# file = open(name4,\"wb\") \n",
    "# pickle.dump(summary_sta2_simlist,file)\n",
    "# file.close()\n",
    "\n",
    "# file = open(name5,\"wb\") \n",
    "# pickle.dump(summary_sta1_inf,file)\n",
    "# file.close()\n",
    "\n",
    "# file = open(name6,\"wb\") \n",
    "# pickle.dump(summary_sta2_inf,file)\n",
    "# file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "637e2ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "        genpoint = [0*5/5.2,400*5/5.2,800*5/5.2,1200*5/5.2,1600*5/5.2,2000*5/5.2,2400*5/5.2,2800*5/5.2,\\\n",
    "                    3200*5/5.2,3600*5/5.2,4000*5/5.2,8000*5/5.2,12000*5/5.2,16000*5/5.2,20000*5/5.2]\n",
    "        inipopsize =[15200*5/5.2, 15200*5/5.2, 15600*5/5.2, 14800*5/5.2, 12400*5/5.2, 10400*5/5.2,9800*5/5.2,\\\n",
    "                     10400*5/5.2,12000*5/5.2,14400*5/5.2,16800*5/5.2,29400*5/5.2,28000*5/5.2,22000*5/5.2,18000*5/5.2]\n",
    "        \n",
    "        \n",
    "        \n",
    "        0"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
